

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Benchmarks &mdash; Medaka 0.11.2 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Variant calling" href="snp.html" />
    <link rel="prev" title="Getting Started" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Medaka
          

          
          </a>

          
            
            
              <div class="version">
                0.11.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Benchmarks</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#comparison-of-medaka-and-nanopolish">Comparison of <cite>medaka</cite> and <cite>nanopolish</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="#evaluation-across-samples-and-depths">Evaluation across samples and depths</a></li>
<li class="toctree-l2"><a class="reference internal" href="#complex-organism-genome-assembly">Complex organism genome assembly</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#assembly-metrics">Assembly metrics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#future-work">Future Work</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="snp.html">Variant calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="methylation.html">Methylation calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html">Walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="draft_origin.html">Origin of the draft sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="future.html">History and Future Directions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Medaka</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Benchmarks</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/benchmarks.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="benchmarks">
<span id="id1"></span><h1>Benchmarks<a class="headerlink" href="#benchmarks" title="Permalink to this headline">¶</a></h1>
<p>The following demonstrates the utility of <code class="docutils literal notranslate"><span class="pre">medaka</span></code>’s neural network in forming an
improved consensus from a pileup of reads.</p>
<p>Results were obtained using the default models provided with <code class="docutils literal notranslate"><span class="pre">medaka</span></code>. These models
were trained using data obtained from a set of prokaroyotic samples.</p>
<p>Error statistics were calculated using the <a class="reference external" href="https://github.com/nanoporetech/pomoxis">pomoxis</a> program <code class="docutils literal notranslate"><span class="pre">assess_assembly</span></code> after
aligning 100kb chunks of the consensus to the reference. Reported metrics are
median values over all chunks.</p>
<div class="section" id="comparison-of-medaka-and-nanopolish">
<h2>Comparison of <cite>medaka</cite> and <cite>nanopolish</cite><a class="headerlink" href="#comparison-of-medaka-and-nanopolish" title="Permalink to this headline">¶</a></h2>
<p>In this comparison R9.4.1 <em>E. coli</em> data from the <a class="reference internal" href="walkthrough.html"><span class="doc">Walkthrough</span></a> were used.
These data were not used to train the model. Basecalling was performed using
<code class="docutils literal notranslate"><span class="pre">Guppy</span> <span class="pre">v3.2.3</span></code>. Basecalled reads were trimmed using <a class="reference external" href="https://github.com/rrwick/Porechop">porechop</a> to remove adapters, and assembly was
performed using <a class="reference external" href="https://github.com/marbl/canu">canu v1.8</a>. The assembly was
optionally corrected using <a class="reference external" href="https://github.com/isovic/racon">racon v1.2.1</a> before being passed
to <code class="docutils literal notranslate"><span class="pre">medaka</span></code> or <code class="docutils literal notranslate"><span class="pre">nanopolish</span></code>. A gpu-enabled version (commit <code class="docutils literal notranslate"><span class="pre">896b8066</span></code>) of
<a class="reference external" href="https://github.com/jts/nanopolish">nanopolish</a> was run from
<a class="reference external" href="https://github.com/jts/nanopolish/pull/661">PR661</a>.</p>
<p>The workflow used here optionally includes a <strong>single</strong> iteration of <code class="docutils literal notranslate"><span class="pre">racon</span></code>, see
<a class="reference internal" href="draft_origin.html#draftorigin"><span class="std std-ref">Origin of the draft sequence</span></a> for further details.</p>
<table class="docutils align-default" id="id2">
<caption><span class="caption-text">Consensus calling of a 90-fold coverage <em>E. coli</em> data set using various methods. <code class="docutils literal notranslate"><span class="pre">Medaka</span></code> is seen
to work effectively, with or without prior application of racon.</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 24%" />
<col style="width: 9%" />
<col style="width: 22%" />
<col style="width: 16%" />
<col style="width: 12%" />
<col style="width: 16%" />
</colgroup>
<tbody>
<tr class="row-odd"><td></td>
<td><p><em>canu</em></p></td>
<td><p><em>medaka no racon</em></p></td>
<td><p><em>racon (x1)</em></p></td>
<td><p><em>medaka</em></p></td>
<td><p><em>nanopolish</em></p></td>
</tr>
<tr class="row-even"><td><p>Q(accuracy)</p></td>
<td><p>25.4</p></td>
<td><p>36.9</p></td>
<td><p>28.8</p></td>
<td><p>37.2</p></td>
<td><p>31.4</p></td>
</tr>
<tr class="row-odd"><td><p>Q(substitution)</p></td>
<td><p>51.0</p></td>
<td><p>53.0</p></td>
<td><p>47.9</p></td>
<td><p>53.0</p></td>
<td><p>47.0</p></td>
</tr>
<tr class="row-even"><td><p>Q(deletion)</p></td>
<td><p>25.4</p></td>
<td><p>37.3</p></td>
<td><p>29.2</p></td>
<td><p>37.7</p></td>
<td><p>31.7</p></td>
</tr>
<tr class="row-odd"><td><p>Q(insertion)</p></td>
<td><p>57.2</p></td>
<td><p>48.2</p></td>
<td><p>41.6</p></td>
<td><p>47.0</p></td>
<td><p>47.0</p></td>
</tr>
<tr class="row-even"><td><p>CPU time / hr:min</p></td>
<td></td>
<td><p>0:05</p></td>
<td><p>0:26</p></td>
<td><p>0:05</p></td>
<td><p>1:54</p></td>
</tr>
</tbody>
</table>
<p>For <code class="docutils literal notranslate"><span class="pre">racon</span></code> and <code class="docutils literal notranslate"><span class="pre">medaka</span></code> the reported CPU times include read to draft
overlapping performed by <code class="docutils literal notranslate"><span class="pre">minimap2</span></code>. <code class="docutils literal notranslate"><span class="pre">medaka</span></code> was run using an
NVIDIA 1080Ti GPU, the total execution time for <code class="docutils literal notranslate"><span class="pre">medaka</span></code> itself was
35 seconds. The wallclock time for <code class="docutils literal notranslate"><span class="pre">nanopolish</span></code> execution was 17 minutes.</p>
<p>A particular advantage of <code class="docutils literal notranslate"><span class="pre">medaka</span></code> over other methods is its improved
accuracy in recovering homopolymer lengths.</p>
<img alt="_images/hp_acc.png" class="align-center" src="_images/hp_acc.png" />
<p>Above the main plot we show homopolymer frequencies from H.sapiens Chrom. 1,
adapted from <a class="reference external" href="http://dirac.cnrs-orleans.fr/~piazza/PB/files/DNA.pdf">Statistical analysis of simple repeats in the human genome</a>.</p>
</div>
<div class="section" id="evaluation-across-samples-and-depths">
<h2>Evaluation across samples and depths<a class="headerlink" href="#evaluation-across-samples-and-depths" title="Permalink to this headline">¶</a></h2>
<p>The comparison below illustrates results at various coverage depths for a
collection of further organisms using both the R9.4.1 and R10 pores. Assemblies were
performed as above with canu and racon, using the <code class="docutils literal notranslate"><span class="pre">Guppy</span> <span class="pre">v3.0.3</span></code> high accuracy
basecaller and <code class="docutils literal notranslate"><span class="pre">medaka</span> <span class="pre">v0.6.5</span></code>.</p>
<img alt="_images/cov_acc.png" class="align-center" src="_images/cov_acc.png" />
</div>
<div class="section" id="complex-organism-genome-assembly">
<h2>Complex organism genome assembly<a class="headerlink" href="#complex-organism-genome-assembly" title="Permalink to this headline">¶</a></h2>
<p>The 100Mbase genome of the nematode worm <em>C. elegans</em> was assembled from a
<a class="reference external" href="https://ont-research.s3-eu-west-1.amazonaws.com/r10_celegans.fq.gz">100-fold coverage R10 dataset</a>,
with median read length of 34kb (all reads &gt;20kb) using the
<a class="reference external" href="https://canu.readthedocs.io/en/latest/index.html">canu</a>,
<a class="reference external" href="https://github.com/fenderglass/Flye">flye</a> and
<a class="reference external" href="https://github.com/chanzuckerberg/shasta">shasta</a> assemblers.
In this experiment, training sets for flip-flop basecallers and medaka models
included a subset of the <em>C. elegans</em> genome. For brevity <code class="docutils literal notranslate"><span class="pre">medaka</span></code> was
applied to drafts direct from the assemblers <em>with no intermediate racon consensus polishing</em>.</p>
<div class="section" id="assembly-metrics">
<h3>Assembly metrics<a class="headerlink" href="#assembly-metrics" title="Permalink to this headline">¶</a></h3>
<p>The table below presents basic assembly metrics for the three assemblers. Run times are
for illustrative purposes only and are not direct comparisons. The assemblers were run
on different hardware appropriate to their predicted run time: <em>canu</em>: a 6 node cluster of
commodity servers, <em>flye</em>: a single server using 20 CPU threads, <em>shasta</em>: a single
server using 46 CPU threads. Assembly metrics were calculated via
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/23422339">quast</a>,
removing contigs less than 5kb in length <a class="reference external" href="https://www.nature.com/articles/s41587-019-0072-8">(Kolmogorov et. al., 2019)</a>.
Accuracy metrics were calculated as described above for the bacterial datasets.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 40%" />
<col style="width: 19%" />
<col style="width: 19%" />
<col style="width: 23%" />
</colgroup>
<tbody>
<tr class="row-odd"><td></td>
<td><p><em>canu</em></p></td>
<td><p><em>flye</em></p></td>
<td><p><em>shasta</em></p></td>
</tr>
<tr class="row-even"><td><p>run time</p></td>
<td><p>~24 hr</p></td>
<td><p>~13 hr</p></td>
<td><p>~34 min</p></td>
</tr>
<tr class="row-odd"><td><p>#contigs</p></td>
<td><p>37</p></td>
<td><p>35</p></td>
<td><p>66</p></td>
</tr>
<tr class="row-even"><td><p>length / Mb</p></td>
<td><p>105</p></td>
<td><p>106</p></td>
<td><p>102</p></td>
</tr>
<tr class="row-odd"><td><p>NG50 / Mb</p></td>
<td><p>5.8</p></td>
<td><p>7</p></td>
<td><p>3.6</p></td>
</tr>
<tr class="row-even"><td><p>NGA50 / Mb</p></td>
<td><p>2.4</p></td>
<td><p>1.8</p></td>
<td><p>1.8</p></td>
</tr>
<tr class="row-odd"><td><p>misassemblies</p></td>
<td><p>221</p></td>
<td><p>163</p></td>
<td><p>93</p></td>
</tr>
<tr class="row-even"><td><p>genome fraction</p></td>
<td><p>99.7%</p></td>
<td><p>94.5%</p></td>
<td><p>99.2%</p></td>
</tr>
<tr class="row-odd"><td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Q(accuracy)</p></td>
<td><p>34.6</p></td>
<td><p>29.2</p></td>
<td><p>32.4</p></td>
</tr>
<tr class="row-odd"><td><p>Q(substitution)</p></td>
<td><p>47.5</p></td>
<td><p>41.2</p></td>
<td><p>44.0</p></td>
</tr>
<tr class="row-even"><td><p>Q(deletion)</p></td>
<td><p>39.2</p></td>
<td><p>36.4</p></td>
<td><p>37.2</p></td>
</tr>
<tr class="row-odd"><td><p>Q(insertion)</p></td>
<td><p>41.1</p></td>
<td><p>33.2</p></td>
<td><p>38.5</p></td>
</tr>
</tbody>
</table>
<p>The results show <code class="docutils literal notranslate"><span class="pre">canu</span></code> to give the overall highest accuracy assembly, though
<code class="docutils literal notranslate"><span class="pre">shasta</span></code> gives similar quality with a fraction of the compute resource. However,
the <code class="docutils literal notranslate"><span class="pre">shasta</span></code> assembly is somewhat more fragmented than the other two assemblies
albeit with a lower misassembly rate.</p>
<p>The accuracy metrics can be calculated also outside of repetitive regions of the
genome by use of <a class="reference external" href="http://www.repeatmasker.org/species/ce.html">RepeatMasker</a>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 40%" />
<col style="width: 19%" />
<col style="width: 19%" />
<col style="width: 23%" />
</colgroup>
<tbody>
<tr class="row-odd"><td></td>
<td><p><em>canu</em></p></td>
<td><p><em>flye</em></p></td>
<td><p><em>shasta</em></p></td>
</tr>
<tr class="row-even"><td><p>Q(accuracy)</p></td>
<td><p>37.3</p></td>
<td><p>32.7</p></td>
<td><p>35.2</p></td>
</tr>
<tr class="row-odd"><td><p>Q(substitution)</p></td>
<td><p>50.0</p></td>
<td><p>43.0</p></td>
<td><p>47.0</p></td>
</tr>
<tr class="row-even"><td><p>Q(deletion)</p></td>
<td><p>40.9</p></td>
<td><p>38.0</p></td>
<td><p>39.2</p></td>
</tr>
<tr class="row-odd"><td><p>Q(insertion)</p></td>
<td><p>44.0</p></td>
<td><p>34.9</p></td>
<td><p>40.9</p></td>
</tr>
</tbody>
</table>
<p>Finally <a class="reference external" href="https://busco.ezlab.org">busco</a> can be used
to measure the gene content of the three assemblies. Results are reported in the usual
“busco notation”, C: complete, S: complete and single copy, D: complete and duplicated,
F: fragmented, M: missing. To aid comparison we also give scores for the reference sequence
used for the assessments above.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 18%" />
<col style="width: 82%" />
</colgroup>
<tbody>
<tr class="row-odd"><td></td>
<td><p>BUSCO scores</p></td>
</tr>
<tr class="row-even"><td><p>reference</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">C:98.6%[S:98.0%,D:0.6%],F:0.8%,M:0.6%,n:982</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>canu</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">C:99.3%[S:98.6%,D:0.7%],F:0.6%,M:0.1%,n:982</span></code></p></td>
</tr>
<tr class="row-even"><td><p>flye</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">C:98.5%[S:98.1%,D:0.4%],F:1.1%,M:0.4%,n:982</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>shasta</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">C:99.0%[S:98.5%,D:0.5%],F:0.9%,M:0.1%,n:982</span></code></p></td>
</tr>
</tbody>
</table>
<p>Curiously it is found that the canu and shasta assemblies show a higher complete count than
the reference sequence (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/assembly/GCF_000002985.6/">WBcel235</a>).</p>
</div>
<div class="section" id="future-work">
<h3>Future Work<a class="headerlink" href="#future-work" title="Permalink to this headline">¶</a></h3>
<p>We have illustrated <code class="docutils literal notranslate"><span class="pre">medaka</span></code>’s role in efficiently creating quality consensus
sequences, and the role of long nanopore reads in creating highly contiguous assemblies
with modest compute requirements. The assemblies are in agreement with recent studies
from <a class="reference external" href="https://genome.cshlp.org/content/12/5/669.full.html">Tyson et. al.</a> and
<a class="reference external" href="https://genome.cshlp.org/content/29/6/1009.full">Yoshimura et. al.</a> in suggesting
that the <em>C. elegans</em> reference has missing sequence. The assemblies above are not the
final word on assembly with nanopore sequencing data; in contrast to latter preceding
reference we have expended relatively little effort in the assembly process, relying
entirely on off-the-shelf automated methods.</p>
<p>Aside from improving assembly methodology we are exploring improved chemistries,
basecallers, and consensus algorithms. Our current algorithmic focus is on so-called
run-length encoded (RLE) methods for both basecalling and consensus. Early work
in this area has shown promise in further reducing homopolymer error in particular.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="snp.html" class="btn btn-neutral float-right" title="Variant calling" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral float-left" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-19, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>