

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>medaka package &mdash; Medaka 0.11.2 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="History and Future Directions" href="future.html" />
    <link rel="prev" title="Development" href="development.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Medaka
          

          
          </a>

          
            
            
              <div class="version">
                0.11.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="benchmarks.html">Benchmarks</a></li>
<li class="toctree-l1"><a class="reference internal" href="snp.html">Variant calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="methylation.html">Methylation calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html">Walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="draft_origin.html">Origin of the draft sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="future.html">History and Future Directions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Medaka</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="development.html">Development</a> &raquo;</li>
        
      <li>medaka package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/medaka.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="medaka-package">
<h1>medaka package<a class="headerlink" href="#medaka-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-medaka.common">
<span id="medaka-common-module"></span><h2>medaka.common module<a class="headerlink" href="#module-medaka.common" title="Permalink to this headline">¶</a></h2>
<p>Commonly used data structures and functions.</p>
<dl class="class">
<dt id="medaka.common.ComprAlignPos">
<em class="property">class </em><code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">ComprAlignPos</code><span class="sig-paren">(</span><em class="sig-param">qpos</em>, <em class="sig-param">qbase</em>, <em class="sig-param">qlen</em>, <em class="sig-param">rpos</em>, <em class="sig-param">rbase</em>, <em class="sig-param">rlen</em><span class="sig-paren">)</span><a class="headerlink" href="#medaka.common.ComprAlignPos" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="method">
<dt id="medaka.common.ComprAlignPos.qbase">
<em class="property">property </em><code class="sig-name descname">qbase</code><a class="headerlink" href="#medaka.common.ComprAlignPos.qbase" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.ComprAlignPos.qlen">
<em class="property">property </em><code class="sig-name descname">qlen</code><a class="headerlink" href="#medaka.common.ComprAlignPos.qlen" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.ComprAlignPos.qpos">
<em class="property">property </em><code class="sig-name descname">qpos</code><a class="headerlink" href="#medaka.common.ComprAlignPos.qpos" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.ComprAlignPos.rbase">
<em class="property">property </em><code class="sig-name descname">rbase</code><a class="headerlink" href="#medaka.common.ComprAlignPos.rbase" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 4</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.ComprAlignPos.rlen">
<em class="property">property </em><code class="sig-name descname">rlen</code><a class="headerlink" href="#medaka.common.ComprAlignPos.rlen" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 5</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.ComprAlignPos.rpos">
<em class="property">property </em><code class="sig-name descname">rpos</code><a class="headerlink" href="#medaka.common.ComprAlignPos.rpos" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="medaka.common.OverlapException">
<em class="property">exception </em><code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">OverlapException</code><a class="reference internal" href="_modules/medaka/common.html#OverlapException"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.OverlapException" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Exception class used when examining range overlaps.</p>
</dd></dl>

<dl class="class">
<dt id="medaka.common.Region">
<em class="property">class </em><code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">Region</code><a class="reference internal" href="_modules/medaka/common.html#Region"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Region" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.common.Region" title="medaka.common.Region"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.common.Region</span></code></a></p>
<p>Represents a genomic region.</p>
<dl class="method">
<dt id="medaka.common.Region.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Region.from_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Region.from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse region string into <cite>Region</cite> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – region str</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Region</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;Ecoli&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">Region</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_name</span><span class="o">=</span><span class="s1">&#39;Ecoli&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Region</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;Ecoli:1000-2000&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">Region</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_name</span><span class="o">=</span><span class="s1">&#39;Ecoli&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Region</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;Ecoli:1000&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">Region</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_name</span><span class="o">=</span><span class="s1">&#39;Ecoli&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Region</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;Ecoli:-1000&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">Region</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_name</span><span class="o">=</span><span class="s1">&#39;Ecoli&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Region</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;Ecoli:500-&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">Region</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_name</span><span class="o">=</span><span class="s1">&#39;Ecoli&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Region</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s1">&#39;A:B:c:500-&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">Region</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">ref_name</span><span class="o">=</span><span class="s1">&#39;A:B:c&#39;</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Region.name">
<em class="property">property </em><code class="sig-name descname">name</code><a class="headerlink" href="#medaka.common.Region.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Samtools-style region string, zero-base end exclusive.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Region.overlaps">
<code class="sig-name descname">overlaps</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Region.overlaps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Region.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine if a region overlaps another.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> – a second Region to test overlap.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if regions overlap.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Region.size">
<em class="property">property </em><code class="sig-name descname">size</code><a class="headerlink" href="#medaka.common.Region.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return size of region.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Region.split">
<code class="sig-name descname">split</code><span class="sig-paren">(</span><em class="sig-param">size</em>, <em class="sig-param">overlap=0</em>, <em class="sig-param">fixed_size=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Region.split"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Region.split" title="Permalink to this definition">¶</a></dt>
<dd><p>Split region into sub-regions of a given length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size</strong> – size of sub-regions.</p></li>
<li><p><strong>overlap</strong> – overlap between ends of sub-regions.</p></li>
<li><p><strong>fixed_size</strong> – ensure all sub-regions are equal in size. If <cite>False</cite>
then the final chunk will be created as the smallest size to
conform with <cite>overlap</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a list of sub-regions.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.common.Relationship">
<em class="property">class </em><code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">Relationship</code><a class="reference internal" href="_modules/medaka/common.html#Relationship"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Relationship" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">enum.Enum</span></code></p>
<p>Enumeration of types of overlap.</p>
<dl class="attribute">
<dt id="medaka.common.Relationship.different_ref_name">
<code class="sig-name descname">different_ref_name</code><em class="property"> = 'Samples come from different reference contigs.'</em><a class="headerlink" href="#medaka.common.Relationship.different_ref_name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.forward_abutted">
<code class="sig-name descname">forward_abutted</code><em class="property"> = 'The end of s1 abuts the start of s2.'</em><a class="headerlink" href="#medaka.common.Relationship.forward_abutted" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.forward_gapped">
<code class="sig-name descname">forward_gapped</code><em class="property"> = 's2 follows s1 with a gab inbetween.'</em><a class="headerlink" href="#medaka.common.Relationship.forward_gapped" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.forward_overlap">
<code class="sig-name descname">forward_overlap</code><em class="property"> = 'The end of s1 overlaps the start of s2.'</em><a class="headerlink" href="#medaka.common.Relationship.forward_overlap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.reverse_abutted">
<code class="sig-name descname">reverse_abutted</code><em class="property"> = 'The end of s2 abuts the start of s1.'</em><a class="headerlink" href="#medaka.common.Relationship.reverse_abutted" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.reverse_gapped">
<code class="sig-name descname">reverse_gapped</code><em class="property"> = 's1 follows s2 with a gab inbetween.'</em><a class="headerlink" href="#medaka.common.Relationship.reverse_gapped" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.reverse_overlap">
<code class="sig-name descname">reverse_overlap</code><em class="property"> = 'The end of s2 overlaps the start of s1.'</em><a class="headerlink" href="#medaka.common.Relationship.reverse_overlap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.s1_within_s2">
<code class="sig-name descname">s1_within_s2</code><em class="property"> = 's1 is fully contained within s2.'</em><a class="headerlink" href="#medaka.common.Relationship.s1_within_s2" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="medaka.common.Relationship.s2_within_s1">
<code class="sig-name descname">s2_within_s1</code><em class="property"> = 's2 is fully contained within s1.'</em><a class="headerlink" href="#medaka.common.Relationship.s2_within_s1" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.common.Sample">
<em class="property">class </em><code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">Sample</code><a class="reference internal" href="_modules/medaka/common.html#Sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.common.Sample" title="medaka.common.Sample"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.common.Sample</span></code></a></p>
<p>Represents a pileup range.</p>
<dl class="method">
<dt id="medaka.common.Sample.amend">
<code class="sig-name descname">amend</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.amend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.amend" title="Permalink to this definition">¶</a></dt>
<dd><p>Create new <cite>Sample</cite> with some attributes changed.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.chunks">
<code class="sig-name descname">chunks</code><span class="sig-paren">(</span><em class="sig-param">chunk_len=1000</em>, <em class="sig-param">overlap=200</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.chunks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.chunks" title="Permalink to this definition">¶</a></dt>
<dd><p>Create overlapping chunks of self.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chunk_len</strong> – chunk length (number of columns)</p></li>
<li><p><strong>overlap</strong> – overlap length.</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>chunked <cite>Sample</cite> instances.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.decode_sample_name">
<em class="property">static </em><code class="sig-name descname">decode_sample_name</code><span class="sig-paren">(</span><em class="sig-param">name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.decode_sample_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.decode_sample_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Decode a the result of Sample.name into a dict.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> – <cite>Sample</cite> object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dict.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.first_pos">
<em class="property">property </em><code class="sig-name descname">first_pos</code><a class="headerlink" href="#medaka.common.Sample.first_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>Zero-based first reference co-ordinate.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.from_samples">
<em class="property">static </em><code class="sig-name descname">from_samples</code><span class="sig-paren">(</span><em class="sig-param">samples</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.from_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.from_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a sample by concatenating an iterable of <cite>Sample</cite> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>samples</strong> – iterable of <cite>Sample</cite> objects.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>Sample</cite> obj</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.is_empty">
<em class="property">property </em><code class="sig-name descname">is_empty</code><a class="headerlink" href="#medaka.common.Sample.is_empty" title="Permalink to this definition">¶</a></dt>
<dd><p>Is pileup empty, synomymous to <cite>sample.size == 0</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.last_pos">
<em class="property">property </em><code class="sig-name descname">last_pos</code><a class="headerlink" href="#medaka.common.Sample.last_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>Zero-based (end inclusive) last reference co-ordinate.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.name">
<em class="property">property </em><code class="sig-name descname">name</code><a class="headerlink" href="#medaka.common.Sample.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Create zero-based (end inclusive) samtools-style region string.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.overlap_indices">
<em class="property">static </em><code class="sig-name descname">overlap_indices</code><span class="sig-paren">(</span><em class="sig-param">s1</em>, <em class="sig-param">s2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.overlap_indices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.overlap_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Find indices by which to trim samples to allow concatenation.</p>
<dl class="simple">
<dt>::</dt><dd><p>#          | end1
s1 …………
s2      ……………
#          | start2</p>
</dd>
</dl>
<p>For example:</p>
<p>would join the samples without gaps or overlap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>s1</strong> – First <cite>Sample</cite> object.</p></li>
<li><p><strong>s2</strong> – Second <cite>Sample</cite> object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(end1, start2</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><cite>OverlapException</cite> if samples do not overlap nor abut.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.relative_position">
<em class="property">static </em><code class="sig-name descname">relative_position</code><span class="sig-paren">(</span><em class="sig-param">s1</em>, <em class="sig-param">s2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.relative_position"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.relative_position" title="Permalink to this definition">¶</a></dt>
<dd><p>Check the relative position of two samples in genomic coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>s2</strong> (<em>s1</em><em>,</em>) – <cite>medaka.common.Sample</cite> objs.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>Relationship</cite> enum member.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.size">
<em class="property">property </em><code class="sig-name descname">size</code><a class="headerlink" href="#medaka.common.Sample.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of columns of pileup.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.slice">
<code class="sig-name descname">slice</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#Sample.slice"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.Sample.slice" title="Permalink to this definition">¶</a></dt>
<dd><p>Slice fields along the genomic axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> – slice or index (on array indices)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>Sample</cite> obj with views of slices of the original <cite>Sample</cite>.</p>
</dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
<span class="gp">... </span>    <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">)],</span>
<span class="gp">... </span>    <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;major&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;minor&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">feat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pos</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="n">Sample</span><span class="p">(</span><span class="s1">&#39;contig1&#39;</span><span class="p">,</span> <span class="n">feat</span> <span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>  
<span class="go">Sample(...features=2, ..., positions=(1, 1), label_probs=None)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span><span class="nb">slice</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span> 
<span class="go">Sample(..., features=array([1, 2]),...)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="medaka.common.Sample.span">
<em class="property">property </em><code class="sig-name descname">span</code><a class="headerlink" href="#medaka.common.Sample.span" title="Permalink to this definition">¶</a></dt>
<dd><p>Size of sample in terms of reference positions.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="medaka.common.get_named_logger">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">get_named_logger</code><span class="sig-paren">(</span><em class="sig-param">name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#get_named_logger"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.get_named_logger" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a logger with a name.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.common.get_regions">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">get_regions</code><span class="sig-paren">(</span><em class="sig-param">bam</em>, <em class="sig-param">region_strs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#get_regions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.get_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Create <cite>Region</cite> objects from a bam and region strings.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>bam</strong> – <cite>.bam</cite> file.</p></li>
<li><p><strong>region_strs</strong> – iterable of str in zero-based (samtools-like)
region format e.g. ref:start-end or filepath containing a
region string per line.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of <cite>Region</cite> objects.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.common.grouper">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">grouper</code><span class="sig-paren">(</span><em class="sig-param">gen</em>, <em class="sig-param">batch_size=4</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#grouper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.grouper" title="Permalink to this definition">¶</a></dt>
<dd><p>Group together elements of an iterable without padding remainder.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.common.initialise_alignment">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">initialise_alignment</code><span class="sig-paren">(</span><em class="sig-param">query_name</em>, <em class="sig-param">reference_id</em>, <em class="sig-param">reference_start</em>, <em class="sig-param">query_sequence</em>, <em class="sig-param">cigarstring</em>, <em class="sig-param">flag</em>, <em class="sig-param">mapping_quality=60</em>, <em class="sig-param">query_qualities=None</em>, <em class="sig-param">tags=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#initialise_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.initialise_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a <cite>Pysam.AlignedSegment</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_name</strong> – name of the query sequence</p></li>
<li><p><strong>reference_id</strong> – index to the reference name</p></li>
<li><p><strong>reference_start</strong> – 0-based index of first leftmost reference
coordinate</p></li>
<li><p><strong>query_sequence</strong> – read sequence bases, including those soft clipped</p></li>
<li><p><strong>cigarstring</strong> – cigar string representing the alignment of query
and reference</p></li>
<li><p><strong>flag</strong> – bitwise flag representing some properties of the alignment
(see SAM format)</p></li>
<li><p><strong>mapping_quality</strong> – optional quality of the mapping or query to
reference</p></li>
<li><p><strong>query_qualities</strong> – optional base qualities of the query, including
soft-clipped ones!</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>pysam.AlignedSegment</cite> object</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.common.loose_version_sort">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">loose_version_sort</code><span class="sig-paren">(</span><em class="sig-param">it</em>, <em class="sig-param">key=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#loose_version_sort"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.loose_version_sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort an iterable.</p>
<p>Items will be sorted with <cite>distutils.version.LooseVersion</cite>,
falling back to regular sort when this fails.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">loose_version_sort</span><span class="p">([</span><span class="s1">&#39;chr10&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">])</span>
<span class="go">[&#39;chr1&#39;, &#39;chr2&#39;, &#39;chr10&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">([</span><span class="s1">&#39;chr10&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">])</span>
<span class="go">[&#39;chr1&#39;, &#39;chr10&#39;, &#39;chr2&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">loose_version_sort</span><span class="p">([</span>
<span class="gp">... </span>    <span class="s1">&#39;chr</span><span class="si">{}</span><span class="s1">c</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span>
<span class="gp">... </span>       <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span> <span class="p">,</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">2</span><span class="p">])])</span>  
<span class="go">[&#39;chr1c1&#39;, &#39;chr1c2&#39;, &#39;chr1c10&#39;, &#39;chr2c1&#39;, ..., &#39;chr10c2&#39;, &#39;chr10c10&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="medaka.common.mkdir_p">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">mkdir_p</code><span class="sig-paren">(</span><em class="sig-param">path</em>, <em class="sig-param">info=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#mkdir_p"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.mkdir_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a directory if it doesn’t exist.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.common.print_data_path">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">print_data_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#print_data_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.print_data_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Print data directory containing models.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.common.read_key_value_tsv">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">read_key_value_tsv</code><span class="sig-paren">(</span><em class="sig-param">fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#read_key_value_tsv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.read_key_value_tsv" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a dictionary from a .tsv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fname</strong> – a .tsv file with two columns: keys and values.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dictionary.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.common.ref_name_from_region_str">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">ref_name_from_region_str</code><span class="sig-paren">(</span><em class="sig-param">region_str</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#ref_name_from_region_str"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.ref_name_from_region_str" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse region strings, returning a list of reference names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – iterable of region strings.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>tuple of reference name str.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.common.reverse_complement">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">reverse_complement</code><span class="sig-paren">(</span><em class="sig-param">seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#reverse_complement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.reverse_complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse complement sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Param</dt>
<dd class="field-odd"><p>input sequence string.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>reverse-complemented string.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.common.sliding_window">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">sliding_window</code><span class="sig-paren">(</span><em class="sig-param">a</em>, <em class="sig-param">window=3</em>, <em class="sig-param">step=1</em>, <em class="sig-param">axis=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#sliding_window"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.sliding_window" title="Permalink to this definition">¶</a></dt>
<dd><p>Sliding window across an array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> – input array.</p></li>
<li><p><strong>window</strong> – window length.</p></li>
<li><p><strong>step</strong> – step length between consecutive windows.</p></li>
<li><p><strong>axis</strong> – axis over which to apply window.</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>windows of the input array.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.common.yield_from_bed">
<code class="sig-prename descclassname">medaka.common.</code><code class="sig-name descname">yield_from_bed</code><span class="sig-paren">(</span><em class="sig-param">bedfile</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/common.html#yield_from_bed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.common.yield_from_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Yield chrom, start, stop tuples from a bed file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bedfile</strong> – str, filepath.</p>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>(str chrom, int start, int stop).</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-medaka.datastore">
<span id="medaka-datastore-module"></span><h2>medaka.datastore module<a class="headerlink" href="#module-medaka.datastore" title="Permalink to this headline">¶</a></h2>
<p>Storing of training and inference data to file.</p>
<dl class="class">
<dt id="medaka.datastore.DataIndex">
<em class="property">class </em><code class="sig-prename descclassname">medaka.datastore.</code><code class="sig-name descname">DataIndex</code><span class="sig-paren">(</span><em class="sig-param">filenames</em>, <em class="sig-param">threads=4</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataIndex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Index and serve samples from multiple <cite>DataStore</cite> compatible files.</p>
<dl class="method">
<dt id="medaka.datastore.DataIndex.index">
<em class="property">property </em><code class="sig-name descname">index</code><a class="headerlink" href="#medaka.datastore.DataIndex.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary describing all samples.</p>
<p>The dictionary maps sample names to their file and HDF group. It is
sorted by reference coordinate.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataIndex.yield_from_feature_files">
<code class="sig-name descname">yield_from_feature_files</code><span class="sig-paren">(</span><em class="sig-param">regions=None</em>, <em class="sig-param">samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataIndex.yield_from_feature_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataIndex.yield_from_feature_files" title="Permalink to this definition">¶</a></dt>
<dd><p>Yield <cite>medaka.common.Sample</cite> objects from one or more feature files.</p>
<dl class="field-list simple">
<dt class="field-odd">Regions</dt>
<dd class="field-odd"><p>list of <cite>medaka.common.Region</cite> s for which to yield samples.</p>
</dd>
<dt class="field-even">Samples</dt>
<dd class="field-even"><p>iterable of sample names to yield (in order in which they
are supplied).</p>
</dd>
<dt class="field-odd">Yields</dt>
<dd class="field-odd"><p><cite>medaka.common.Sample</cite> objects.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.datastore.DataStore">
<em class="property">class </em><code class="sig-prename descclassname">medaka.datastore.</code><code class="sig-name descname">DataStore</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">mode='r'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Read and write data to .hdf files.</p>
<dl class="method">
<dt id="medaka.datastore.DataStore.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore.close"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close filehandle of back-end file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.copy_meta">
<code class="sig-name descname">copy_meta</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore.copy_meta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore.copy_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy metadata to another file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> – filename of another file.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.get_meta">
<code class="sig-name descname">get_meta</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore.get_meta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore.get_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>Load (deserialise) a meta data item.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> – name of item to load.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.load_sample">
<code class="sig-name descname">load_sample</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore.load_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore.load_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Load <cite>medaka.common.Sample</cite> object from file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> – str, sample name.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>medaka.common.Sample</cite> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.n_samples">
<em class="property">property </em><code class="sig-name descname">n_samples</code><a class="headerlink" href="#medaka.datastore.DataStore.n_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of samples stored in file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.sample_registry">
<em class="property">property </em><code class="sig-name descname">sample_registry</code><a class="headerlink" href="#medaka.datastore.DataStore.sample_registry" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a set of samples stored within file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.set_meta">
<code class="sig-name descname">set_meta</code><span class="sig-paren">(</span><em class="sig-param">obj</em>, <em class="sig-param">key</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore.set_meta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore.set_meta" title="Permalink to this definition">¶</a></dt>
<dd><p>Store (serialize) a meta data item to file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obj</strong> – the object to serialise.</p></li>
<li><p><strong>key</strong> – the name of the object.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.datastore.DataStore.write_sample">
<code class="sig-name descname">write_sample</code><span class="sig-paren">(</span><em class="sig-param">sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/datastore.html#DataStore.write_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.datastore.DataStore.write_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Write sample to hdf.</p>
<p>Checks are performed to ensure a sample is not written twice and
a count of unique training labels seen is maintained.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample</strong> – <cite>medaka.common.Sample</cite> object.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-medaka.features">
<span id="medaka-features-module"></span><h2>medaka.features module<a class="headerlink" href="#module-medaka.features" title="Permalink to this headline">¶</a></h2>
<p>Creation of neural network input features.</p>
<dl class="class">
<dt id="medaka.features.BaseFeatureEncoder">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">BaseFeatureEncoder</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#BaseFeatureEncoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.BaseFeatureEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for creation of feature arrays from a <cite>.bam</cite> file.</p>
<dl class="method">
<dt id="medaka.features.BaseFeatureEncoder.bam_to_sample">
<code class="sig-name descname">bam_to_sample</code><span class="sig-paren">(</span><em class="sig-param">reads_bam</em>, <em class="sig-param">region</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#BaseFeatureEncoder.bam_to_sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.BaseFeatureEncoder.bam_to_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a section of an alignment pileup to a sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reads_bam</strong> – (sorted indexed) bam with read alignment to reference</p></li>
<li><p><strong>region</strong> – <cite>medaka.common.Region</cite> object with ref_name, start and
end attributes.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>medaka.common.Sample</cite> object</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.features.BaseFeatureEncoder.bams_to_training_samples">
<em class="property">abstract </em><code class="sig-name descname">bams_to_training_samples</code><span class="sig-paren">(</span><em class="sig-param">truth_bam</em>, <em class="sig-param">bam</em>, <em class="sig-param">region</em>, <em class="sig-param">label_scheme</em>, <em class="sig-param">truth_haplotag=None</em>, <em class="sig-param">min_length=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#BaseFeatureEncoder.bams_to_training_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.BaseFeatureEncoder.bams_to_training_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Create labelled samples, should internally call bam_to_sample.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.features.BaseFeatureEncoder.feature_vector_length">
<em class="property">abstract property </em><code class="sig-name descname">feature_vector_length</code><a class="headerlink" href="#medaka.features.BaseFeatureEncoder.feature_vector_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Return size of a single feature sample.</p>
<p>The length of a neural network input at a single time point.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.features.CountsFeatureEncoder">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">CountsFeatureEncoder</code><span class="sig-paren">(</span><em class="sig-param">normalise='total'</em>, <em class="sig-param">dtypes=(''</em>, <em class="sig-param">)</em>, <em class="sig-param">tag_name=None</em>, <em class="sig-param">tag_value=None</em>, <em class="sig-param">tag_keep_missing=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#CountsFeatureEncoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.CountsFeatureEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.features.BaseFeatureEncoder" title="medaka.features.BaseFeatureEncoder"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.features.BaseFeatureEncoder</span></code></a></p>
<p>Create a pileup array of counts of observed bases.</p>
<dl class="method">
<dt id="medaka.features.CountsFeatureEncoder.bams_to_training_samples">
<code class="sig-name descname">bams_to_training_samples</code><span class="sig-paren">(</span><em class="sig-param">truth_bam</em>, <em class="sig-param">bam</em>, <em class="sig-param">region</em>, <em class="sig-param">label_scheme</em>, <em class="sig-param">truth_haplotag=None</em>, <em class="sig-param">min_length=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#CountsFeatureEncoder.bams_to_training_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.CountsFeatureEncoder.bams_to_training_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare training data chunks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>truth_bam</strong> – .bam file of truth aligned to ref to generate labels.</p></li>
<li><p><strong>bam</strong> – input .bam file.</p></li>
<li><p><strong>region</strong> – <cite>medaka.common.Region</cite> instance for region to process.</p></li>
<li><p><strong>label_scheme</strong> – a <cite>LabelScheme</cite> describing network outputs.</p></li>
<li><p><strong>truth_haplotag</strong> – two letter tag name used for grouping truth
labels by haplotype.</p></li>
<li><p><strong>min_length</strong> – minimum length for valid alignments.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>tuple of <cite>medaka.common.Sample</cite> objects.</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Chunks might be missing if <cite>truth_bam</cite> is provided and
regions with multiple mappings were encountered.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="medaka.features.CountsFeatureEncoder.feature_dtype">
<code class="sig-name descname">feature_dtype</code><a class="headerlink" href="#medaka.features.CountsFeatureEncoder.feature_dtype" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.float32</span></code></p>
</dd></dl>

<dl class="method">
<dt id="medaka.features.CountsFeatureEncoder.feature_vector_length">
<em class="property">property </em><code class="sig-name descname">feature_vector_length</code><a class="headerlink" href="#medaka.features.CountsFeatureEncoder.feature_vector_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of a single sample.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.features.FeatureEncoderMeta">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">FeatureEncoderMeta</code><a class="reference internal" href="_modules/medaka/features.html#FeatureEncoderMeta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.FeatureEncoderMeta" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code>, <a class="reference internal" href="#medaka.features.FeatureEncoderRegistrar" title="medaka.features.FeatureEncoderRegistrar"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.features.FeatureEncoderRegistrar</span></code></a></p>
<p>Metaclass facilitating registration of <cite>FeatureEncoder</cite> s.</p>
</dd></dl>

<dl class="class">
<dt id="medaka.features.FeatureEncoderRegistrar">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">FeatureEncoderRegistrar</code><a class="reference internal" href="_modules/medaka/features.html#FeatureEncoderRegistrar"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.FeatureEncoderRegistrar" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">type</span></code></p>
<p>Class for registering feature encoders.</p>
<dl class="method">
<dt id="medaka.features.FeatureEncoderRegistrar.register_feature_encoder">
<code class="sig-name descname">register_feature_encoder</code><span class="sig-paren">(</span><em class="sig-param">cls</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#FeatureEncoderRegistrar.register_feature_encoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.FeatureEncoderRegistrar.register_feature_encoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Add <cite>FeatureEncoder</cite> to <cite>feature_encoders</cite> dict.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.features.HardRLEFeatureEncoder">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">HardRLEFeatureEncoder</code><span class="sig-paren">(</span><em class="sig-param">normalise='total'</em>, <em class="sig-param">dtypes=(''</em>, <em class="sig-param">)</em>, <em class="sig-param">tag_name=None</em>, <em class="sig-param">tag_value=None</em>, <em class="sig-param">tag_keep_missing=False</em>, <em class="sig-param">num_qstrat=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#HardRLEFeatureEncoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.HardRLEFeatureEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.features.CountsFeatureEncoder" title="medaka.features.CountsFeatureEncoder"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.features.CountsFeatureEncoder</span></code></a></p>
<p>Create a pileup array of counts of observed bases.</p>
<p>Counts are segregated according to run lengths stored in the quality
information of reads.</p>
<dl class="method">
<dt id="medaka.features.HardRLEFeatureEncoder.feature_vector_length">
<em class="property">property </em><code class="sig-name descname">feature_vector_length</code><a class="headerlink" href="#medaka.features.HardRLEFeatureEncoder.feature_vector_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of a single sample.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.features.SampleGenerator">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">SampleGenerator</code><span class="sig-paren">(</span><em class="sig-param">bam</em>, <em class="sig-param">region</em>, <em class="sig-param">feature_encoder</em>, <em class="sig-param">truth_bam=None</em>, <em class="sig-param">label_scheme=None</em>, <em class="sig-param">truth_haplotag=None</em>, <em class="sig-param">chunk_len=1000</em>, <em class="sig-param">chunk_overlap=200</em>, <em class="sig-param">enable_chunking=True</em>, <em class="sig-param">min_truth_length=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#SampleGenerator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.SampleGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Orchestration of neural network inputs with chunking.</p>
<dl class="method">
<dt id="medaka.features.SampleGenerator.samples">
<em class="property">property </em><code class="sig-name descname">samples</code><a class="headerlink" href="#medaka.features.SampleGenerator.samples" title="Permalink to this definition">¶</a></dt>
<dd><p>List of (possibly) chunked samples.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.features.SoftRLEFeatureEncoder">
<em class="property">class </em><code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">SoftRLEFeatureEncoder</code><span class="sig-paren">(</span><em class="sig-param">normalise='total'</em>, <em class="sig-param">dtypes=(''</em>, <em class="sig-param">)</em>, <em class="sig-param">tag_name=None</em>, <em class="sig-param">tag_value=None</em>, <em class="sig-param">tag_keep_missing=False</em>, <em class="sig-param">num_qstrat=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#SoftRLEFeatureEncoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.SoftRLEFeatureEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.features.HardRLEFeatureEncoder" title="medaka.features.HardRLEFeatureEncoder"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.features.HardRLEFeatureEncoder</span></code></a></p>
<p>Create pileups using soft RLE calls.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.features.create_samples">
<code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">create_samples</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#create_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.create_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for creation of feature .hdfs (labelled or unlabelled).</p>
</dd></dl>

<dl class="function">
<dt id="medaka.features.pileup_counts">
<code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">pileup_counts</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bam</em>, <em class="sig-param">dtype_prefixes=None</em>, <em class="sig-param">region_split=100000</em>, <em class="sig-param">workers=8</em>, <em class="sig-param">tag_name=None</em>, <em class="sig-param">tag_value=None</em>, <em class="sig-param">keep_missing=False</em>, <em class="sig-param">num_qstrat=1</em>, <em class="sig-param">weibull_summation=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#pileup_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.pileup_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Create pileup counts feature array for region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – <cite>medaka.common.Region</cite> object</p></li>
<li><p><strong>bam</strong> – .bam file with alignments.</p></li>
<li><p><strong>dtype_prefixes</strong> – prefixes for query names which to separate counts.
If <cite>None</cite> (or of length 1), counts are not split.</p></li>
<li><p><strong>region_split</strong> – largest region to process in single thread.</p></li>
<li><p><strong>workers</strong> – worker threads for calculating pileup.</p></li>
<li><p><strong>tag_name</strong> – two letter tag name by which to filter reads.</p></li>
<li><p><strong>tag_value</strong> – integer value of tag for reads to keep.</p></li>
<li><p><strong>keep_missing</strong> – whether to keep reads when tag is missing.</p></li>
<li><p><strong>num_qstrat</strong> – number of layers for qscore stratification.</p></li>
<li><p><strong>weibull_summation</strong> – use a Weibull partial-counts approach,
requires ‘WL’ and ‘WK’ float-array tags.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pileup counts array, reference positions, insertion postions</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.features.pileup_counts_norm_indices">
<code class="sig-prename descclassname">medaka.features.</code><code class="sig-name descname">pileup_counts_norm_indices</code><span class="sig-paren">(</span><em class="sig-param">dtypes</em>, <em class="sig-param">num_qstrat=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/features.html#pileup_counts_norm_indices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.features.pileup_counts_norm_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate feature vector normalization groups.</p>
<p>Calculates (per-datatype, per-read-orientation) locations of bases in
<cite>pileup_counts</cite> output for the purpose of various normalization
strategies.</p>
<p>e.g. For two datatype the feature vector is counts of bases:
(acgtACGTdDacgtACGTdD), i.e. the first datatype followed by
the second, and where lowercase denotes reverse reads and the
base ordering is defined in the library consts <cite>plp_bases</cite>. The
resultant dictionary contains entries such as:
(datatype1, False):[4,5,6,7,9], for the datatype1 forward bases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dtypes</strong> – list of datatype names.</p></li>
<li><p><strong>num_qstrat</strong> – number of layers for qscore stratification.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a dictionary of the form
<cite>{(datatype, is_rev): [base1_index, base2_index, …]}</cite>.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-medaka.keras_ext">
<span id="medaka-keras-ext-module"></span><h2>medaka.keras_ext module<a class="headerlink" href="#module-medaka.keras_ext" title="Permalink to this headline">¶</a></h2>
<p>Extensions to keras API for medaka.</p>
<dl class="class">
<dt id="medaka.keras_ext.ModelMetaCheckpoint">
<em class="property">class </em><code class="sig-prename descclassname">medaka.keras_ext.</code><code class="sig-name descname">ModelMetaCheckpoint</code><span class="sig-paren">(</span><em class="sig-param">medaka_meta</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#ModelMetaCheckpoint"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.ModelMetaCheckpoint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tensorflow.python.keras.callbacks.ModelCheckpoint</span></code></p>
<p>Custom ModelCheckpoint to add medaka-specific metadata.</p>
<dl class="method">
<dt id="medaka.keras_ext.ModelMetaCheckpoint.on_epoch_end">
<code class="sig-name descname">on_epoch_end</code><span class="sig-paren">(</span><em class="sig-param">epoch</em>, <em class="sig-param">logs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#ModelMetaCheckpoint.on_epoch_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.ModelMetaCheckpoint.on_epoch_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform actions at the end of an epoch.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.keras_ext.SequenceBatcher">
<em class="property">class </em><code class="sig-prename descclassname">medaka.keras_ext.</code><code class="sig-name descname">SequenceBatcher</code><span class="sig-paren">(</span><em class="sig-param">batcher</em>, <em class="sig-param">dataset='train'</em>, <em class="sig-param">mini_epochs=1</em>, <em class="sig-param">seed=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#SequenceBatcher"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.SequenceBatcher" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tensorflow.python.keras.utils.data_utils.Sequence</span></code></p>
<p>Interface for keras to a <cite>TrainBatcher</cite> for training and validation.</p>
<dl class="method">
<dt id="medaka.keras_ext.SequenceBatcher.on_epoch_end">
<code class="sig-name descname">on_epoch_end</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#SequenceBatcher.on_epoch_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.SequenceBatcher.on_epoch_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform actions at the end of an epoch.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.keras_ext.TrainValTensorBoard">
<em class="property">class </em><code class="sig-prename descclassname">medaka.keras_ext.</code><code class="sig-name descname">TrainValTensorBoard</code><span class="sig-paren">(</span><em class="sig-param">log_dir='./logs'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#TrainValTensorBoard"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.TrainValTensorBoard" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tensorflow.python.keras.callbacks_v1.TensorBoard</span></code></p>
<p>Modification of tensorboard to plot test and validation together.</p>
<dl class="method">
<dt id="medaka.keras_ext.TrainValTensorBoard.on_epoch_end">
<code class="sig-name descname">on_epoch_end</code><span class="sig-paren">(</span><em class="sig-param">epoch</em>, <em class="sig-param">logs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#TrainValTensorBoard.on_epoch_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.TrainValTensorBoard.on_epoch_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Write logs on epoch end.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.keras_ext.TrainValTensorBoard.on_train_end">
<code class="sig-name descname">on_train_end</code><span class="sig-paren">(</span><em class="sig-param">logs=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#TrainValTensorBoard.on_train_end"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.TrainValTensorBoard.on_train_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Close the validation writer on exit.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.keras_ext.TrainValTensorBoard.set_model">
<code class="sig-name descname">set_model</code><span class="sig-paren">(</span><em class="sig-param">model</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/keras_ext.html#TrainValTensorBoard.set_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.keras_ext.TrainValTensorBoard.set_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Set writer for validation metrics.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-medaka.labels">
<span id="medaka-labels-module"></span><h2>medaka.labels module<a class="headerlink" href="#module-medaka.labels" title="Permalink to this headline">¶</a></h2>
<p>Schemes to encode and decode truth labels and network outputs.</p>
<dl class="class">
<dt id="medaka.labels.BaseLabelScheme">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">BaseLabelScheme</code><a class="reference internal" href="_modules/medaka/labels.html#BaseLabelScheme"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.BaseLabelScheme" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Base class for labelling schemes.</p>
<p>Labels are ‘truth’ or ‘y’ vectors used to train models.
Models aim to predict a label (y) from a feature (X).</p>
<p>A LabelScheme contains all logic for creating label vectors
from a source of truth (typically an alignment of true reference
sequence(s) to a draft sequence).</p>
<p>A LabelScheme class also contains all logic for converting the
model output into useful output (usually  a consensus
sequence or a set of variant calls).</p>
<p>Encoding (conversion of truth alignments to training vectors):</p>
<p>The public method <cite>self.encode</cite> converts alignments
to arrays of integer-encoded intermediate representations that
are saved to file during feature creation.
These private method must be implemented.</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>_alignment_to_pairs</cite> (called by _alignment_to_labels)</p></li>
<li><p><cite>_labels_to_encoded_labels</cite></p></li>
</ul>
</div></blockquote>
<p>At training time, it is necessary to call the following public
method for the conversion of intermediate forms saved to file to
training vectors input to the model training process.</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>encoded_labels_to_training_vectors</cite></p></li>
</ul>
</div></blockquote>
<p>Decoding (interpretation of network outputs):</p>
<p>For SNP calling, the public method is <cite>self.decode_snps</cite>
This private method must be implemented.</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>_prob_to_snp</cite> (for SNP decoding)</p></li>
</ul>
</div></blockquote>
<p>Public methods may also be defined for other types of decoding.</p>
<p>For example, the concrete HaploidLabelScheme defines:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>decode_consensus</cite></p></li>
<li><p><cite>decode_variants</cite></p></li>
</ul>
</div></blockquote>
<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.decode_snps">
<code class="sig-name descname">decode_snps</code><span class="sig-paren">(</span><em class="sig-param">sample</em>, <em class="sig-param">ref_seq</em>, <em class="sig-param">ref_vcf=None</em>, <em class="sig-param">threshold=0.04</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#BaseLabelScheme.decode_snps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.BaseLabelScheme.decode_snps" title="Permalink to this definition">¶</a></dt>
<dd><p>Decode network outputs to medaka.vcf.Variant objects recording SNPs.</p>
<p>Optionally, SNPs are returned at ALL loci specified in suplied ref_vcf.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample</strong> – <cite>medaka.common.Sample</cite></p></li>
<li><p><strong>ref_seq</strong> – reference sequence, should be upper case.</p></li>
<li><p><strong>ref_vcf</strong> – <cite>.vcf</cite> file</p></li>
<li><p><strong>threshold</strong> – threshold for acceptance of secondary call.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of <cite>medaka.vcf.Variant</cite> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.encode">
<code class="sig-name descname">encode</code><span class="sig-paren">(</span><em class="sig-param">truth_alns</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#BaseLabelScheme.encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.BaseLabelScheme.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert truth alignment(s) to array of intermediate representation.</p>
<p>In most cases the intermediate representation consists of integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>truth_alns</strong> – tuple of <cite>pysam.AlignedSegment</cite> s for each haplotype
spanning the same genomic range.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>tuple(positions, training_vectors)</p>
<ul class="simple">
<li><p>positions: numpy structured array with ‘major’
(reference position index) and ‘minor’
(trailing insertion index) fields.</p></li>
<li><p>encoded: nd.array of encoded labels</p></li>
</ul>
</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is generally the case that the returned encoded labels must be
padded with encoded gap labels when aligned to corresponding
training feature data.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.encoded_labels_to_training_vectors">
<em class="property">abstract </em><code class="sig-name descname">encoded_labels_to_training_vectors</code><span class="sig-paren">(</span><em class="sig-param">enc_labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#BaseLabelScheme.encoded_labels_to_training_vectors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.BaseLabelScheme.encoded_labels_to_training_vectors" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert integer (tuple) encoded labels to truth vectors.</p>
<p>(e.g. one-hot encoded truth vectors) that represent the truth
for comparison with network output logits in
e.g. metric(truth, pred) or loss(truth, pred)) functions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>enc_labels</strong> – np.ndarray of integer (tuples)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>np.ndarray of training vectors</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.n_elements">
<em class="property">abstract property </em><code class="sig-name descname">n_elements</code><a class="headerlink" href="#medaka.labels.BaseLabelScheme.n_elements" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements provided by truth alignment.</p>
<p>(Mostly) synonymous with ploidy. i.e. n_elements = 2 where
two symbols are provided for the two haplotypes of a truth
alignment for diploid training.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.num_classes">
<em class="property">abstract property </em><code class="sig-name descname">num_classes</code><a class="headerlink" href="#medaka.labels.BaseLabelScheme.num_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements in output layer of neural network.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.padding_vector">
<em class="property">abstract property </em><code class="sig-name descname">padding_vector</code><a class="headerlink" href="#medaka.labels.BaseLabelScheme.padding_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the integer encoding used to denote a gap.</p>
<p>Used to inform calling programs that need to expand
label arrays to align with feature arrays,
where reads introduce minor positions.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.BaseLabelScheme.snp_metainfo">
<em class="property">property </em><code class="sig-name descname">snp_metainfo</code><a class="headerlink" href="#medaka.labels.BaseLabelScheme.snp_metainfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Return meta data for use in <cite>.vcf</cite> header.</p>
</dd></dl>

<dl class="attribute">
<dt id="medaka.labels.BaseLabelScheme.symbols">
<code class="sig-name descname">symbols</code><em class="property"> = '*ACGT'</em><a class="headerlink" href="#medaka.labels.BaseLabelScheme.symbols" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.labels.DiploidLabelScheme">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">DiploidLabelScheme</code><a class="reference internal" href="_modules/medaka/labels.html#DiploidLabelScheme"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.DiploidLabelScheme" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.labels.BaseLabelScheme" title="medaka.labels.BaseLabelScheme"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.labels.BaseLabelScheme</span></code></a></p>
<p>LabelScheme defines a two-element label per genomic position.</p>
<p>Each label maps to an integer, and we make a direct diploid call.
A categorical crossentropy loss is appropriate.</p>
<p>Encoding: (‘A’, ‘C’) -&gt; 6 -&gt; [6] (a sparse one-hot encoded training vector)</p>
<p>SNP decoding: [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0] -&gt; (‘A’, ‘C’)
(network output is not constrained to be 0 or 1, but for simplicity
we show the argmax probability as 1 and all others as 0)</p>
<p>Depending on the reference, this may result in a homozygous or
heterozygous SNP call.</p>
<p>Where reference is (‘A’,), we output ref=’A’, alt=[‘C’], GT=’0/1’
Where reference is (‘T’,), we output ref=’T’ alt=[‘A’, ‘C’], GT=’1/2’</p>
<p>When predicion is homozygous e.g. (‘A’, ‘A’):
Where reference is (‘T’,), we output ref=’T’, alt=[‘A’], GT=’1/1’</p>
<p>When the prediction contains a gap symbol e.g. (‘A’, ‘*’):
Where reference is (‘T’,), we output ref=’T’, alt=[‘A’], GT=’1/1’</p>
<dl class="method">
<dt id="medaka.labels.DiploidLabelScheme.encoded_labels_to_training_vectors">
<code class="sig-name descname">encoded_labels_to_training_vectors</code><span class="sig-paren">(</span><em class="sig-param">enc_labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#DiploidLabelScheme.encoded_labels_to_training_vectors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.DiploidLabelScheme.encoded_labels_to_training_vectors" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert integer (tuple) encoded labels to truth vectors.</p>
<p>(e.g. sparse one-hot encoded truth vectors) that represent the truth
for comparison with network output logits in
e.g. metric(truth, pred) or loss(truth, pred)) functions.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidLabelScheme.n_elements">
<em class="property">property </em><code class="sig-name descname">n_elements</code><a class="headerlink" href="#medaka.labels.DiploidLabelScheme.n_elements" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements provided by truth alignment.</p>
<p>Synonymous with ploidy.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidLabelScheme.num_classes">
<em class="property">property </em><code class="sig-name descname">num_classes</code><a class="headerlink" href="#medaka.labels.DiploidLabelScheme.num_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements in output layer of neural network.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidLabelScheme.padding_vector">
<em class="property">property </em><code class="sig-name descname">padding_vector</code><a class="headerlink" href="#medaka.labels.DiploidLabelScheme.padding_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the integer encoding used to denote a gap.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidLabelScheme.snp_metainfo">
<em class="property">property </em><code class="sig-name descname">snp_metainfo</code><a class="headerlink" href="#medaka.labels.DiploidLabelScheme.snp_metainfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Return meta data for use in <cite>.vcf</cite> header.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.labels.DiploidZygosityLabelScheme">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">DiploidZygosityLabelScheme</code><a class="reference internal" href="_modules/medaka/labels.html#DiploidZygosityLabelScheme"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.DiploidZygosityLabelScheme" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.labels.DiploidLabelScheme" title="medaka.labels.DiploidLabelScheme"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.labels.DiploidLabelScheme</span></code></a></p>
<p>Diploid label scheme with an explicit heterozygosity element.</p>
<p>The intermediate integer encoding is identical to the
DiploidLabelScheme.</p>
<p>The training vector encoding is multi-hot. For example, to encode
(‘A’, C’) at a locus, three elements of the training vectors representing
‘A’, ‘C’ and ‘is_heterogygous’ are set to 1.</p>
<p>(‘A’, ‘C’) -&gt; 6 -&gt; [0, 1, 1, 0, 0, 1]</p>
<p>During inference, we predict the probability of each base existing
independently, and also (independently) the probability of the locus
being heterozygous. A binary crossentropy loss is appropriate.
The output vector will not sum to 1.</p>
<p>Importantly, during inference, there may be an is_heterozygous
prediction &gt; 0.5 which conflicts with independent predictions of which
(how many) symbols are present (which symbols have independent
probabilities &gt; 0.5. In this case, the is_heterozygous call dominates
and we return the top two most probable bases.</p>
<p>SNP decoding:</p>
<p>[0, 0.6, 0.7, 0, 0, 0.8] -&gt; (‘C’, ‘A’) (in order of most probable)</p>
<p>Depending on the reference, this may result in a homozygous or
heterozygous SNP call.</p>
<p>Where reference is (‘A’,), we output ref=’A’, alt=[‘C’], GT=’0/1’
Where reference is (‘T’,), we output ref=’T’ alt=[‘A’, ‘C’], GT=’1/2’</p>
<p>When predicion is homozygous e.g. (‘A’, ‘A’):
Where reference is (‘T’,), we output ref=’T’, alt=[‘A’], GT=’1/1’</p>
<p>When the prediction contains a gap symbol e.g. (‘A’, ‘*’):
Where reference is (‘T’,), we output ref=’T’, alt=[‘A’], GT=’1/1’</p>
<p>If the is_heterozygous had been &lt; 0.5, this would alter the call.
[0, 0,6, 0.7, 0, 0, 0.4] -&gt; (‘C’, ‘C’)</p>
<dl class="method">
<dt id="medaka.labels.DiploidZygosityLabelScheme.encoded_labels_to_training_vectors">
<code class="sig-name descname">encoded_labels_to_training_vectors</code><span class="sig-paren">(</span><em class="sig-param">enc_labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#DiploidZygosityLabelScheme.encoded_labels_to_training_vectors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.DiploidZygosityLabelScheme.encoded_labels_to_training_vectors" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert integer (tuple) encoded labels to truth vectors.</p>
<p>(e.g. one-hot encoded truth vectors) that represent the truth
for comparison with network output logits in
e.g. metric(truth, pred) or loss(truth, pred)) functions.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidZygosityLabelScheme.n_elements">
<em class="property">property </em><code class="sig-name descname">n_elements</code><a class="headerlink" href="#medaka.labels.DiploidZygosityLabelScheme.n_elements" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements provided by truth alignment.</p>
<p>Synonymous with ploidy.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidZygosityLabelScheme.num_classes">
<em class="property">property </em><code class="sig-name descname">num_classes</code><a class="headerlink" href="#medaka.labels.DiploidZygosityLabelScheme.num_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements in output layer of neural network.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.DiploidZygosityLabelScheme.padding_vector">
<em class="property">property </em><code class="sig-name descname">padding_vector</code><a class="headerlink" href="#medaka.labels.DiploidZygosityLabelScheme.padding_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the integer encoding used to denote a gap.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.labels.HaploidLabelScheme">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">HaploidLabelScheme</code><a class="reference internal" href="_modules/medaka/labels.html#HaploidLabelScheme"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.HaploidLabelScheme" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.labels.BaseLabelScheme" title="medaka.labels.BaseLabelScheme"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.labels.BaseLabelScheme</span></code></a></p>
<p>A single-element label per genomic position.</p>
<p>Unitary symbols map to integers, which map to one-hot encoded
training vectors.</p>
<p>Encoding (‘A’,) -&gt; 1 -&gt; [0, 1, 0, 0, 0]</p>
<p>Consensus decoding is simple argmax.
[[0.02, 0.9, 0.02, 0.01, 0.05], [0.01, 0.02, 0.01, 0.9, 0.05]] -&gt; “AG”</p>
<p>SNP decoding utilises a hard threshold to define secondary calls.</p>
<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.decode_consensus">
<code class="sig-name descname">decode_consensus</code><span class="sig-paren">(</span><em class="sig-param">sample</em>, <em class="sig-param">with_gaps=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#HaploidLabelScheme.decode_consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.decode_consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert network output to consensus sequence by argmax decoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample</strong> – medaka.common.Sample</p></li>
<li><p><strong>with_gaps</strong> – include gap (“*”) characters in output.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str, consensus sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.decode_variants">
<code class="sig-name descname">decode_variants</code><span class="sig-paren">(</span><em class="sig-param">sample</em>, <em class="sig-param">ref_seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#HaploidLabelScheme.decode_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.decode_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert network output in sample to a set of <cite>medaka.vcf.Variant</cite> s.</p>
<p>A consensus sequence is decoded and compared with a reference sequence.
Both substitution and indel variants that may be multi-base will be
reported in the output <cite>medaka.vcf.Variant</cite> s.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample</strong> – <cite>medaka.common.Sample</cite>.</p></li>
<li><p><strong>ref_seq</strong> – reference sequence, should be upper case.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of <cite>medaka.vcf.Variant</cite> objects.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.encoded_labels_to_training_vectors">
<code class="sig-name descname">encoded_labels_to_training_vectors</code><span class="sig-paren">(</span><em class="sig-param">enc_labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#HaploidLabelScheme.encoded_labels_to_training_vectors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.encoded_labels_to_training_vectors" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert integer (tuple) encoded labels to truth vectors.</p>
<p>(e.g. sparse one-hot encoded truth vectors) that represent the truth
for comparison with network output logits in
e.g. metric(truth, pred) or loss(truth, pred)) functions.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.n_elements">
<em class="property">property </em><code class="sig-name descname">n_elements</code><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.n_elements" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements provided by truth alignment.</p>
<p>Synonymous with ploidy.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.num_classes">
<em class="property">property </em><code class="sig-name descname">num_classes</code><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.num_classes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return number of elements in output layer of neural network.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.padding_vector">
<em class="property">property </em><code class="sig-name descname">padding_vector</code><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.padding_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the integer encoding used to denote a gap.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.HaploidLabelScheme.variant_metainfo">
<em class="property">property </em><code class="sig-name descname">variant_metainfo</code><a class="headerlink" href="#medaka.labels.HaploidLabelScheme.variant_metainfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Return meta data for use in <cite>.vcf</cite> header.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.labels.LabelSchemeMeta">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">LabelSchemeMeta</code><a class="reference internal" href="_modules/medaka/labels.html#LabelSchemeMeta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.LabelSchemeMeta" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code>, <a class="reference internal" href="#medaka.labels.LabelSchemeRegistrar" title="medaka.labels.LabelSchemeRegistrar"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.labels.LabelSchemeRegistrar</span></code></a></p>
<p>Metaclass facilitating registration of <cite>LabelScheme</cite> s.</p>
</dd></dl>

<dl class="class">
<dt id="medaka.labels.LabelSchemeRegistrar">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">LabelSchemeRegistrar</code><a class="reference internal" href="_modules/medaka/labels.html#LabelSchemeRegistrar"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.LabelSchemeRegistrar" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">type</span></code></p>
<p>Class for registering label schemes.</p>
<dl class="method">
<dt id="medaka.labels.LabelSchemeRegistrar.register_label_scheme">
<code class="sig-name descname">register_label_scheme</code><span class="sig-paren">(</span><em class="sig-param">cls</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#LabelSchemeRegistrar.register_label_scheme"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.LabelSchemeRegistrar.register_label_scheme" title="Permalink to this definition">¶</a></dt>
<dd><p>Add <cite>LabelScheme</cite> to <cite>label_schemes</cite> dict.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.labels.RLELabelScheme">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">RLELabelScheme</code><span class="sig-paren">(</span><em class="sig-param">max_run=12</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#RLELabelScheme"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.RLELabelScheme" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#medaka.labels.HaploidLabelScheme" title="medaka.labels.HaploidLabelScheme"><code class="xref py py-class docutils literal notranslate"><span class="pre">medaka.labels.HaploidLabelScheme</span></code></a></p>
<p>Class for RLE labelling schemes.</p>
<p>The true length of the runs is encoded in the query scores.</p>
<dl class="method">
<dt id="medaka.labels.RLELabelScheme.decode_consensus">
<code class="sig-name descname">decode_consensus</code><span class="sig-paren">(</span><em class="sig-param">sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#RLELabelScheme.decode_consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.RLELabelScheme.decode_consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert network output to consensus sequence by argmax decoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample</strong> – medaka.common.Sample</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str, consensus sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.labels.RLELabelScheme.padding_vector">
<em class="property">property </em><code class="sig-name descname">padding_vector</code><a class="headerlink" href="#medaka.labels.RLELabelScheme.padding_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the integer encoding used to denote a gap.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.labels.TruthAlignment">
<em class="property">class </em><code class="sig-prename descclassname">medaka.labels.</code><code class="sig-name descname">TruthAlignment</code><span class="sig-paren">(</span><em class="sig-param">alignment</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#TruthAlignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.TruthAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Process truth alignments.</p>
<dl class="method">
<dt id="medaka.labels.TruthAlignment.bam_to_alignments">
<em class="property">static </em><code class="sig-name descname">bam_to_alignments</code><span class="sig-paren">(</span><em class="sig-param">truth_bam</em>, <em class="sig-param">region</em>, <em class="sig-param">haplotag=None</em>, <em class="sig-param">min_length=1000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/labels.html#TruthAlignment.bam_to_alignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.labels.TruthAlignment.bam_to_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Get processed truth alignments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>truth_bam</strong> – (sorted indexed) bam with true sequence
aligned to reference</p></li>
<li><p><strong>region</strong> – <cite>medaka.common.Region</cite> obj,
(all alignments with any overlap with the interval
start:end will be retrieved)</p></li>
<li><p><strong>haplotag</strong> – bam tag specifying which haplotype the alignment
belongs to (for polyploid labels)</p></li>
<li><p><strong>min_length</strong> – minimum length for valid alignments.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of tuples where each tuple contains <cite>TruthAlignment</cite>
for each haplotype trimmed to common genomic window.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-medaka.medaka">
<span id="medaka-medaka-module"></span><h2>medaka.medaka module<a class="headerlink" href="#module-medaka.medaka" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="medaka.medaka.CheckBam">
<em class="property">class </em><code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">CheckBam</code><span class="sig-paren">(</span><em class="sig-param">option_strings</em>, <em class="sig-param">dest</em>, <em class="sig-param">nargs=None</em>, <em class="sig-param">const=None</em>, <em class="sig-param">default=None</em>, <em class="sig-param">type=None</em>, <em class="sig-param">choices=None</em>, <em class="sig-param">required=False</em>, <em class="sig-param">help=None</em>, <em class="sig-param">metavar=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#CheckBam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.CheckBam" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.Action</span></code></p>
<p>Check a bam has &lt; 2 samples (RG tag)</p>
</dd></dl>

<dl class="class">
<dt id="medaka.medaka.CheckIsBed">
<em class="property">class </em><code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">CheckIsBed</code><span class="sig-paren">(</span><em class="sig-param">option_strings</em>, <em class="sig-param">dest</em>, <em class="sig-param">nargs=None</em>, <em class="sig-param">const=None</em>, <em class="sig-param">default=None</em>, <em class="sig-param">type=None</em>, <em class="sig-param">choices=None</em>, <em class="sig-param">required=False</em>, <em class="sig-param">help=None</em>, <em class="sig-param">metavar=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#CheckIsBed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.CheckIsBed" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.Action</span></code></p>
<p>Check if –region option is a bed file.</p>
</dd></dl>

<dl class="class">
<dt id="medaka.medaka.ResolveModel">
<em class="property">class </em><code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">ResolveModel</code><span class="sig-paren">(</span><em class="sig-param">option_strings</em>, <em class="sig-param">dest</em>, <em class="sig-param">default=None</em>, <em class="sig-param">required=False</em>, <em class="sig-param">help='Model file.'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#ResolveModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.ResolveModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.Action</span></code></p>
<p>Resolve model filename or ID into filename</p>
</dd></dl>

<dl class="class">
<dt id="medaka.medaka.StoreDict">
<em class="property">class </em><code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">StoreDict</code><span class="sig-paren">(</span><em class="sig-param">option_strings</em>, <em class="sig-param">dest</em>, <em class="sig-param">nargs=None</em>, <em class="sig-param">const=None</em>, <em class="sig-param">default=None</em>, <em class="sig-param">type=None</em>, <em class="sig-param">choices=None</em>, <em class="sig-param">required=False</em>, <em class="sig-param">help=None</em>, <em class="sig-param">metavar=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#StoreDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.StoreDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.Action</span></code></p>
</dd></dl>

<dl class="function">
<dt id="medaka.medaka.fastrle">
<code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">fastrle</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#fastrle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.fastrle" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="medaka.medaka.is_rle_encoder">
<code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">is_rle_encoder</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#is_rle_encoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.is_rle_encoder" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="medaka.medaka.main">
<code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="medaka.medaka.print_all_models">
<code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">print_all_models</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#print_all_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.print_all_models" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="medaka.medaka.print_model_path">
<code class="sig-prename descclassname">medaka.medaka.</code><code class="sig-name descname">print_model_path</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka.html#print_model_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka.print_model_path" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-medaka.medaka_counts">
<span id="medaka-medaka-counts-module"></span><h2>medaka.medaka_counts module<a class="headerlink" href="#module-medaka.medaka_counts" title="Permalink to this headline">¶</a></h2>
<p>Simple test/demonstration program of pileup counting.</p>
<dl class="function">
<dt id="medaka.medaka_counts.main">
<code class="sig-prename descclassname">medaka.medaka_counts.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/medaka_counts.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.medaka_counts.main" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for pileup counting program.</p>
</dd></dl>

</div>
<div class="section" id="module-medaka.methdaka">
<span id="medaka-methdaka-module"></span><h2>medaka.methdaka module<a class="headerlink" href="#module-medaka.methdaka" title="Permalink to this headline">¶</a></h2>
<p>Functionality for aggregating single read methylation calls.</p>
<dl class="function">
<dt id="medaka.methdaka.call_methylation">
<code class="sig-prename descclassname">medaka.methdaka.</code><code class="sig-name descname">call_methylation</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/methdaka.html#call_methylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.methdaka.call_methylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for calling methylation from bam file.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.methdaka.hdf_to_sam">
<code class="sig-prename descclassname">medaka.methdaka.</code><code class="sig-name descname">hdf_to_sam</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/methdaka.html#hdf_to_sam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.methdaka.hdf_to_sam" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for converting guppy methylcalled fast5s to sam.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.methdaka.hdf_to_sam_worker">
<code class="sig-prename descclassname">medaka.methdaka.</code><code class="sig-name descname">hdf_to_sam_worker</code><span class="sig-paren">(</span><em class="sig-param">reference</em>, <em class="sig-param">fname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/methdaka.html#hdf_to_sam_worker"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.methdaka.hdf_to_sam_worker" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract and align basecall and methylation data from <cite>.fast5</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reference</strong> – <cite>.fasta</cite> file containing reference sequence(s).</p></li>
<li><p><strong>fname</strong> – <cite>.’fast5</cite> file containing read data.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-medaka.models">
<span id="medaka-models-module"></span><h2>medaka.models module<a class="headerlink" href="#module-medaka.models" title="Permalink to this headline">¶</a></h2>
<p>Creation and loading of models.</p>
<dl class="function">
<dt id="medaka.models.build_majority">
<code class="sig-prename descclassname">medaka.models.</code><code class="sig-name descname">build_majority</code><span class="sig-paren">(</span><em class="sig-param">feature_len</em>, <em class="sig-param">num_classes</em>, <em class="sig-param">gru_size=128</em>, <em class="sig-param">classify_activation='softmax'</em>, <em class="sig-param">time_steps=None</em>, <em class="sig-param">allow_cudnn=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/models.html#build_majority"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.models.build_majority" title="Permalink to this definition">¶</a></dt>
<dd><p>Build a mock model that simply sums counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>feature_len</strong> – int, number of features for each pileup column.</p></li>
<li><p><strong>num_classes</strong> – int, number of output class labels.</p></li>
<li><p><strong>gru_size</strong> – int, size of each GRU layer.</p></li>
<li><p><strong>classify_activation</strong> – str, activation to use in classification layer.</p></li>
<li><p><strong>time_steps</strong> – int, number of pileup columns in a sample.</p></li>
<li><p><strong>allow_cudnn</strong> – bool, opt-in to cudnn when using a GPU.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>keras.models.Sequential</cite> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.models.build_model">
<code class="sig-prename descclassname">medaka.models.</code><code class="sig-name descname">build_model</code><span class="sig-paren">(</span><em class="sig-param">feature_len</em>, <em class="sig-param">num_classes</em>, <em class="sig-param">gru_size=128</em>, <em class="sig-param">classify_activation='softmax'</em>, <em class="sig-param">time_steps=None</em>, <em class="sig-param">allow_cudnn=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/models.html#build_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.models.build_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Build a bidirectional GRU model with CuDNNGRU support.</p>
<p>CuDNNGRU implementation is claimed to give speed-up on GPU of 7x.
The function will build a model capable of running on GPU with
CuDNNGRU provided a) a GPU is present, b) the option has been
allowed by the <cite>allow_cudnn</cite> argument; otherwise a compatible
(but not CuDNNGRU accelerated model) is built.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>feature_len</strong> – int, number of features for each pileup column.</p></li>
<li><p><strong>num_classes</strong> – int, number of output class labels.</p></li>
<li><p><strong>gru_size</strong> – int, size of each GRU layer.</p></li>
<li><p><strong>classify_activation</strong> – str, activation to use in classification layer.</p></li>
<li><p><strong>time_steps</strong> – int, number of pileup columns in a sample.</p></li>
<li><p><strong>allow_cudnn</strong> – bool, opt-in to cudnn when using a GPU.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>keras.models.Sequential</cite> object.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.models.load_model">
<code class="sig-prename descclassname">medaka.models.</code><code class="sig-name descname">load_model</code><span class="sig-paren">(</span><em class="sig-param">fname</em>, <em class="sig-param">time_steps=None</em>, <em class="sig-param">allow_cudnn=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/models.html#load_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.models.load_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a model from an .hdf file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> – .hdf file containing model.</p></li>
<li><p><strong>time_steps</strong> – number of time points in RNN, <cite>None</cite> for dynamic.</p></li>
<li><p><strong>allow_cudnn</strong> – allow use of CuDNN optimizations.</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>..note:: keras’ <cite>load_model</cite> cannot handle CuDNNGRU layers, hence this</dt><dd><p>function builds the model then loads the weights.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-medaka.prediction">
<span id="medaka-prediction-module"></span><h2>medaka.prediction module<a class="headerlink" href="#module-medaka.prediction" title="Permalink to this headline">¶</a></h2>
<p>Inference program and ancilliary functions.</p>
<dl class="class">
<dt id="medaka.prediction.DataLoader">
<em class="property">class </em><code class="sig-prename descclassname">medaka.prediction.</code><code class="sig-name descname">DataLoader</code><span class="sig-paren">(</span><em class="sig-param">sample_cache_size</em>, <em class="sig-param">bam</em>, <em class="sig-param">regions</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/prediction.html#DataLoader"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.prediction.DataLoader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Loading of data for inference.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.prediction.predict">
<code class="sig-prename descclassname">medaka.prediction.</code><code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/prediction.html#predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.prediction.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Inference program.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.prediction.run_prediction">
<code class="sig-prename descclassname">medaka.prediction.</code><code class="sig-name descname">run_prediction</code><span class="sig-paren">(</span><em class="sig-param">output</em>, <em class="sig-param">bam</em>, <em class="sig-param">regions</em>, <em class="sig-param">model</em>, <em class="sig-param">feature_encoder</em>, <em class="sig-param">chunk_len</em>, <em class="sig-param">chunk_ovlp</em>, <em class="sig-param">batch_size=200</em>, <em class="sig-param">save_features=False</em>, <em class="sig-param">enable_chunking=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/prediction.html#run_prediction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.prediction.run_prediction" title="Permalink to this definition">¶</a></dt>
<dd><p>Inference worker.</p>
</dd></dl>

</div>
<div class="section" id="module-medaka.rle">
<span id="medaka-rle-module"></span><h2>medaka.rle module<a class="headerlink" href="#module-medaka.rle" title="Permalink to this headline">¶</a></h2>
<p>Run length encoding and realignment of reads.</p>
<dl class="class">
<dt id="medaka.rle.RLEConverter">
<em class="property">class </em><code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">RLEConverter</code><span class="sig-paren">(</span><em class="sig-param">basecall</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#RLEConverter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.RLEConverter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class to convert a basecall to RLE, with coordinate conversions.</p>
<dl class="method">
<dt id="medaka.rle.RLEConverter.coord_compact_to_full">
<code class="sig-name descname">coord_compact_to_full</code><span class="sig-paren">(</span><em class="sig-param">coord</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#RLEConverter.coord_compact_to_full"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.RLEConverter.coord_compact_to_full" title="Permalink to this definition">¶</a></dt>
<dd><p>Map from encoded index to full basecall index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>coord</strong> – array_like, compact coordinate(s) to be converted to
full basecall</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>numpy.ndarray of coordinates</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.rle.RLEConverter.transform_coords">
<code class="sig-name descname">transform_coords</code><span class="sig-paren">(</span><em class="sig-param">start</em>, <em class="sig-param">end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#RLEConverter.transform_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.RLEConverter.transform_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Transform a slice co-ordinates from the input sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start co-ordinate in uncompressed sequence.</p></li>
<li><p><strong>end</strong> – end co-ordinate in uncompressed sequence (exclusive).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the trimmed compressed sequence.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.rle.RLEConverter.trimmed_compact">
<code class="sig-name descname">trimmed_compact</code><span class="sig-paren">(</span><em class="sig-param">start</em>, <em class="sig-param">end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#RLEConverter.trimmed_compact"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.RLEConverter.trimmed_compact" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a trimmed compressed sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> – start co-ordinate in uncompressed sequence.</p></li>
<li><p><strong>end</strong> – end co-ordinate in uncompressed sequence (exclusive).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the trimmed compressed sequence.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="medaka.rle.add_extra_clipping">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">add_extra_clipping</code><span class="sig-paren">(</span><em class="sig-param">cigar</em>, <em class="sig-param">start_clipped</em>, <em class="sig-param">end_clipped</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#add_extra_clipping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.add_extra_clipping" title="Permalink to this definition">¶</a></dt>
<dd><p>Add extra soft clipping to begining and end of cigar string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cigar</strong> – input cigar string</p></li>
<li><p><strong>start_clipped</strong> – soft clipping to be added at the start of the read</p></li>
<li><p><strong>end_clipped</strong> – soft clipping to be added at the end of the read</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>modified cigar string</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.rle.compress_bam">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">compress_bam</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#compress_bam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.compress_bam" title="Permalink to this definition">¶</a></dt>
<dd><p>Compress a bam alignment file into an RLE system of reference.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.rle.compress_seq">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">compress_seq</code><span class="sig-paren">(</span><em class="sig-param">read</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#compress_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.compress_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Compress homopolymers within a basecall.</p>
<p>Creates and RLE sequence record, with run-lengths stored in qualities.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>read</strong> – <cite>pysam FastxRecord</cite> object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>pysam FastxRecord</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.rle.get_rl_params">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">get_rl_params</code><span class="sig-paren">(</span><em class="sig-param">read_name</em>, <em class="sig-param">read_fast5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#get_rl_params"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.get_rl_params" title="Permalink to this definition">¶</a></dt>
<dd><p>Get shape and scale parameters from fast5 for read.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.rle.parasail_alignment">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">parasail_alignment</code><span class="sig-paren">(</span><em class="sig-param">query</em>, <em class="sig-param">ref</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#parasail_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.parasail_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Run a Smith-Waterman alignment between two sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query</strong> – the query sequence.</p></li>
<li><p><strong>ref</strong> – the reference sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>reference start co-ordinate, cigar string</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.rle.rle">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">rle</code><span class="sig-paren">(</span><em class="sig-param">iterable</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#rle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.rle" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate a run length encoding (rle), of an input iterable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>iterable</strong> – input iterable.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>structured array with fields <cite>start</cite>, <cite>length</cite>, and <cite>value</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.rle.rlebam">
<code class="sig-prename descclassname">medaka.rle.</code><code class="sig-name descname">rlebam</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/rle.html#rlebam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.rle.rlebam" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for merging run length information for fast5s to bam.</p>
</dd></dl>

</div>
<div class="section" id="module-medaka.smolecule">
<span id="medaka-smolecule-module"></span><h2>medaka.smolecule module<a class="headerlink" href="#module-medaka.smolecule" title="Permalink to this headline">¶</a></h2>
<p>Creation of consensus sequences from repetitive reads.</p>
<dl class="class">
<dt id="medaka.smolecule.Alignment">
<em class="property">class </em><code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">Alignment</code><span class="sig-paren">(</span><em class="sig-param">rname</em>, <em class="sig-param">qname</em>, <em class="sig-param">flag</em>, <em class="sig-param">rstart</em>, <em class="sig-param">seq</em>, <em class="sig-param">cigar</em><span class="sig-paren">)</span><a class="headerlink" href="#medaka.smolecule.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="method">
<dt id="medaka.smolecule.Alignment.cigar">
<em class="property">property </em><code class="sig-name descname">cigar</code><a class="headerlink" href="#medaka.smolecule.Alignment.cigar" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 5</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Alignment.flag">
<em class="property">property </em><code class="sig-name descname">flag</code><a class="headerlink" href="#medaka.smolecule.Alignment.flag" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Alignment.qname">
<em class="property">property </em><code class="sig-name descname">qname</code><a class="headerlink" href="#medaka.smolecule.Alignment.qname" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Alignment.rname">
<em class="property">property </em><code class="sig-name descname">rname</code><a class="headerlink" href="#medaka.smolecule.Alignment.rname" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Alignment.rstart">
<em class="property">property </em><code class="sig-name descname">rstart</code><a class="headerlink" href="#medaka.smolecule.Alignment.rstart" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Alignment.seq">
<em class="property">property </em><code class="sig-name descname">seq</code><a class="headerlink" href="#medaka.smolecule.Alignment.seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 4</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.smolecule.Read">
<em class="property">class </em><code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">Read</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">subreads</em>, <em class="sig-param">initialize=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Functionality to extract information from a read with subreads.</p>
<dl class="method">
<dt id="medaka.smolecule.Read.align_to_template">
<code class="sig-name descname">align_to_template</code><span class="sig-paren">(</span><em class="sig-param">template</em>, <em class="sig-param">template_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.align_to_template"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.align_to_template" title="Permalink to this definition">¶</a></dt>
<dd><p>Align subreads to a template sequence using Smith-Waterman.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> – sequence to which to align subreads.</p></li>
<li><p><strong>template_name</strong> – name of template sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>Alignment</cite> tuples.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.from_fastx">
<em class="property">classmethod </em><code class="sig-name descname">from_fastx</code><span class="sig-paren">(</span><em class="sig-param">fastx</em>, <em class="sig-param">name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.from_fastx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.from_fastx" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a Read from a fasta/q file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fastx</strong> – input file path.</p></li>
<li><p><strong>name</strong> – name of Read. If not given the filename will be used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.initialize">
<code class="sig-name descname">initialize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.initialize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.initialize" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate initial alignments of subreads to scaffold read.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.mappy_to_template">
<code class="sig-name descname">mappy_to_template</code><span class="sig-paren">(</span><em class="sig-param">template</em>, <em class="sig-param">template_name</em>, <em class="sig-param">align=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.mappy_to_template"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.mappy_to_template" title="Permalink to this definition">¶</a></dt>
<dd><p>Align subreads to a template sequence using minimap.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>template</strong> – sequence to which to align subreads.</p></li>
<li><p><strong>template_name</strong> – name of template sequence.</p></li>
<li><p><strong>align</strong> – retrieve cigar string (else produce paf)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>Alignment</cite> tuples.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.multi_from_fastx">
<em class="property">classmethod </em><code class="sig-name descname">multi_from_fastx</code><span class="sig-paren">(</span><em class="sig-param">fastx</em>, <em class="sig-param">take_all=False</em>, <em class="sig-param">read_id=None</em>, <em class="sig-param">depth_filter=1</em>, <em class="sig-param">length_filter=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.multi_from_fastx"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.multi_from_fastx" title="Permalink to this definition">¶</a></dt>
<dd><p>Create multiple <cite>Read</cite> s from a fasta/q file.</p>
<p>It is assumed that subreads are grouped by read and named with
&lt;read_id&gt;_&lt;subread_id&gt;.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fastx</strong> – input file path.</p></li>
<li><p><strong>take_all</strong> – skip check on subread_ids, take all subreads in one
<cite>Read</cite>.</p></li>
<li><p><strong>read_id</strong> – name of <cite>Read</cite>. Only used for <cite>take_all == True</cite>. If
not given the basename of the input file is used.</p></li>
<li><p><strong>depth_filter</strong> – require reads to have at least this many subreads.</p></li>
<li><p><strong>length_filter</strong> – require reads to have a median subread length
above this value.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.nseqs">
<em class="property">property </em><code class="sig-name descname">nseqs</code><a class="headerlink" href="#medaka.smolecule.Read.nseqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the number of subreads contained in the read.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.orient_subreads">
<code class="sig-name descname">orient_subreads</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.orient_subreads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.orient_subreads" title="Permalink to this definition">¶</a></dt>
<dd><p>Find orientation of subreads with respect to consensus sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><cite>Alignment</cite> s of subreads to consensus.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.poa_consensus">
<code class="sig-name descname">poa_consensus</code><span class="sig-paren">(</span><em class="sig-param">additional_seq=None</em>, <em class="sig-param">method='racon'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#Read.poa_consensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.Read.poa_consensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a consensus sequence for the read.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Read.seqs">
<em class="property">property </em><code class="sig-name descname">seqs</code><a class="headerlink" href="#medaka.smolecule.Read.seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of the subread sequences.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.smolecule.Subread">
<em class="property">class </em><code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">Subread</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">seq</em><span class="sig-paren">)</span><a class="headerlink" href="#medaka.smolecule.Subread" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="method">
<dt id="medaka.smolecule.Subread.name">
<em class="property">property </em><code class="sig-name descname">name</code><a class="headerlink" href="#medaka.smolecule.Subread.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="medaka.smolecule.Subread.seq">
<em class="property">property </em><code class="sig-name descname">seq</code><a class="headerlink" href="#medaka.smolecule.Subread.seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="medaka.smolecule.first_cigar">
<code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">first_cigar</code><span class="sig-paren">(</span><em class="sig-param">cigar</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#first_cigar"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.first_cigar" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract details of the first operation in a cigar string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cigar</strong> – cigar string.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>op. length, op. type</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.smolecule.main">
<code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.main" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for repeat read consensus creation.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.smolecule.parasail_to_sam">
<code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">parasail_to_sam</code><span class="sig-paren">(</span><em class="sig-param">result</em>, <em class="sig-param">seq</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#parasail_to_sam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.parasail_to_sam" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract reference start and sam compatible cigar string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>result</strong> – parasail alignment result.</p></li>
<li><p><strong>seq</strong> – query sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>reference start coordinate, cigar string.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.smolecule.poa_workflow">
<code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">poa_workflow</code><span class="sig-paren">(</span><em class="sig-param">reads</em>, <em class="sig-param">threads</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#poa_workflow"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.poa_workflow" title="Permalink to this definition">¶</a></dt>
<dd><p>Worker function for processing repetitive reads.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>reads</strong> – list of <cite>Read</cite> s.</p></li>
<li><p><strong>threads</strong> – number of threads to use for processing.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.smolecule.write_bam">
<code class="sig-prename descclassname">medaka.smolecule.</code><code class="sig-name descname">write_bam</code><span class="sig-paren">(</span><em class="sig-param">fname</em>, <em class="sig-param">alignments</em>, <em class="sig-param">header</em>, <em class="sig-param">bam=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/smolecule.html#write_bam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.smolecule.write_bam" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a <cite>.bam</cite> file for a set of alignments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> – output filename.</p></li>
<li><p><strong>alignments</strong> – a list of <cite>Alignment</cite> tuples.</p></li>
<li><p><strong>header</strong> – bam header</p></li>
<li><p><strong>bam</strong> – write bam, else sam</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-medaka.stitch">
<span id="medaka-stitch-module"></span><h2>medaka.stitch module<a class="headerlink" href="#module-medaka.stitch" title="Permalink to this headline">¶</a></h2>
<p>Creation of contiguous consensus sequences from chunked network outputs.</p>
<dl class="function">
<dt id="medaka.stitch.stitch">
<code class="sig-prename descclassname">medaka.stitch.</code><code class="sig-name descname">stitch</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/stitch.html#stitch"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.stitch.stitch" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for stitching program.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.stitch.stitch_from_probs">
<code class="sig-prename descclassname">medaka.stitch.</code><code class="sig-name descname">stitch_from_probs</code><span class="sig-paren">(</span><em class="sig-param">h5_fp</em>, <em class="sig-param">regions=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/stitch.html#stitch_from_probs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.stitch.stitch_from_probs" title="Permalink to this definition">¶</a></dt>
<dd><p>Join overlapping label probabilities from HDF5 files.</p>
<blockquote>
<div><p>Network outputs from multiple samples stored within a file are spliced
together into a logically contiguous array and decoded to generate
contiguous sequence(s).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>h5_fp</strong> – iterable of HDF5 filepaths</p></li>
<li><p><strong>regions</strong> – iterable of region (strings) to process</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of (region string, sequence)</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.stitch.write_fasta">
<code class="sig-prename descclassname">medaka.stitch.</code><code class="sig-name descname">write_fasta</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">contigs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/stitch.html#write_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.stitch.write_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a fasta file from tuples of (name, sequence).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> – output filename.</p></li>
<li><p><strong>contigs</strong> – tuples of the form (sequence name, base sequence).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-medaka.training">
<span id="medaka-training-module"></span><h2>medaka.training module<a class="headerlink" href="#module-medaka.training" title="Permalink to this headline">¶</a></h2>
<p>Training program and ancillary functions.</p>
<dl class="class">
<dt id="medaka.training.TrainBatcher">
<em class="property">class </em><code class="sig-prename descclassname">medaka.training.</code><code class="sig-name descname">TrainBatcher</code><span class="sig-paren">(</span><em class="sig-param">features</em>, <em class="sig-param">validation=0.2</em>, <em class="sig-param">seed=0</em>, <em class="sig-param">batch_size=500</em>, <em class="sig-param">threads=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#TrainBatcher"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.TrainBatcher" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Batching of training and validation samples.</p>
<dl class="method">
<dt id="medaka.training.TrainBatcher.sample_to_x_y">
<code class="sig-name descname">sample_to_x_y</code><span class="sig-paren">(</span><em class="sig-param">sample</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#TrainBatcher.sample_to_x_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.TrainBatcher.sample_to_x_y" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a <cite>common.Sample</cite> object into a training x, y tuple.</p>
<p>This method is synonymous to the static method
sample_to_x_y_bq_worker.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample</strong> – (filename, sample key)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(np.ndarray of inputs, np.ndarray of labels)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.training.TrainBatcher.sample_to_x_y_bq_worker">
<em class="property">static </em><code class="sig-name descname">sample_to_x_y_bq_worker</code><span class="sig-paren">(</span><em class="sig-param">sample</em>, <em class="sig-param">label_scheme</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#TrainBatcher.sample_to_x_y_bq_worker"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.TrainBatcher.sample_to_x_y_bq_worker" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a <cite>common.Sample</cite> object into a training x, y tuple.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample</strong> – (sample key, filename).</p></li>
<li><p><strong>label_scheme</strong> – <cite>LabelScheme</cite> obj.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(np.ndarray of inputs, np.ndarray of labels)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.training.TrainBatcher.samples_to_batch">
<code class="sig-name descname">samples_to_batch</code><span class="sig-paren">(</span><em class="sig-param">samples</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#TrainBatcher.samples_to_batch"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.TrainBatcher.samples_to_batch" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a set of <cite>common.Sample</cite> objects into a training X, Y tuple.</p>
<p>The function wraps <cite>.sample_to_x_y</cite> and stacks the outputs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>samples</strong> – (filename, sample key) tuples</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(np.ndarray of inputs, np.ndarray of labels)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="medaka.training.cat_acc">
<code class="sig-prename descclassname">medaka.training.</code><code class="sig-name descname">cat_acc</code><span class="sig-paren">(</span><em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#cat_acc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.cat_acc" title="Permalink to this definition">¶</a></dt>
<dd><p>Keras loss function for sparse_categorical_accuracy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y_true</strong> – tensor of true class labels.</p></li>
<li><p><strong>y_pred</strong> – class output scores from network.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>categorical accuracy.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.training.qscore">
<code class="sig-prename descclassname">medaka.training.</code><code class="sig-name descname">qscore</code><span class="sig-paren">(</span><em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#qscore"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.qscore" title="Permalink to this definition">¶</a></dt>
<dd><p>Keras metric function for calculating scaled error.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y_true</strong> – tensor of true class labels.</p></li>
<li><p><strong>y_pred</strong> – class output scores from network.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>class error expressed as a phred score.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.training.run_training">
<code class="sig-prename descclassname">medaka.training.</code><code class="sig-name descname">run_training</code><span class="sig-paren">(</span><em class="sig-param">train_name</em>, <em class="sig-param">batcher</em>, <em class="sig-param">model_fp=None</em>, <em class="sig-param">epochs=5000</em>, <em class="sig-param">class_weight=None</em>, <em class="sig-param">n_mini_epochs=1</em>, <em class="sig-param">threads_io=1</em>, <em class="sig-param">optimizer='rmsprop'</em>, <em class="sig-param">optim_args=None</em>, <em class="sig-param">allow_cudnn=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#run_training"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.run_training" title="Permalink to this definition">¶</a></dt>
<dd><p>Run training.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.training.train">
<code class="sig-prename descclassname">medaka.training.</code><code class="sig-name descname">train</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/training.html#train"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.training.train" title="Permalink to this definition">¶</a></dt>
<dd><p>Training program.</p>
</dd></dl>

</div>
<div class="section" id="module-medaka.variant">
<span id="medaka-variant-module"></span><h2>medaka.variant module<a class="headerlink" href="#module-medaka.variant" title="Permalink to this headline">¶</a></h2>
<p>Creation of variant call files from network outputs.</p>
<dl class="function">
<dt id="medaka.variant.join_samples">
<code class="sig-prename descclassname">medaka.variant.</code><code class="sig-name descname">join_samples</code><span class="sig-paren">(</span><em class="sig-param">sample_gen</em>, <em class="sig-param">ref_seq</em>, <em class="sig-param">label_scheme</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/variant.html#join_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.variant.join_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Process stream of trimmed Samples.</p>
<p>Care is taken to ensure a variant is not split across
multiple Samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_gen</strong> – stream of (<cite>medaka.common.Sample</cite>,
bool is_last_in_contig)</p></li>
<li><p><strong>ref_seq</strong> – str, reference sequence</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><cite>medaka.common.Sample</cite> s</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.variant.snps_from_hdf">
<code class="sig-prename descclassname">medaka.variant.</code><code class="sig-name descname">snps_from_hdf</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/variant.html#snps_from_hdf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.variant.snps_from_hdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for SNP calling from HDF5 files.</p>
<p>A <cite>LabelScheme</cite> read from file is used to decode SNPs. All <cite>LabelScheme</cite> s
define a <cite>decode_snps</cite> public method. We do not need to use <cite>join_samples</cite>
to look for variants overlapping sample slice boundaries because we only
analyse a single locus at a time. This means that <cite>LabelScheme</cite> s that do
not define the <cite>decode_consensus</cite> method (called within <cite>join_samples</cite>) can
be used.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.variant.trim_samples">
<code class="sig-prename descclassname">medaka.variant.</code><code class="sig-name descname">trim_samples</code><span class="sig-paren">(</span><em class="sig-param">sample_gen</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/variant.html#trim_samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.variant.trim_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate trimmed samples.</p>
<p>Samples are trimmed to remove overlap between adjacent samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_gen</strong> – generator yielding <cite>medaka.common.Sample</cite> s</p>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p>(<cite>medaka.common.Sample</cite> view, bool is_last_in_contig)</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.variant.variants_from_hdf">
<code class="sig-prename descclassname">medaka.variant.</code><code class="sig-name descname">variants_from_hdf</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/variant.html#variants_from_hdf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.variant.variants_from_hdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for variant calling from HDF5 files.</p>
<p>A <cite>LabelScheme</cite> read from HDF must define both a <cite>decode_variants</cite>
and <cite>decode_consnesus</cite> method. The latter is used with <cite>join_samples</cite>
to detect multi-locus variants spanning <cite>Sample</cite> slice boundaries.</p>
</dd></dl>

</div>
<div class="section" id="module-medaka.vcf">
<span id="medaka-vcf-module"></span><h2>medaka.vcf module<a class="headerlink" href="#module-medaka.vcf" title="Permalink to this headline">¶</a></h2>
<p>Reading and writing of Variant Call Format files.</p>
<dl class="class">
<dt id="medaka.vcf.Haploid2DiploidConverter">
<em class="property">class </em><code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">Haploid2DiploidConverter</code><span class="sig-paren">(</span><em class="sig-param">vcf1</em>, <em class="sig-param">vcf2</em>, <em class="sig-param">ref_fasta</em>, <em class="sig-param">only_overlapping=True</em>, <em class="sig-param">discard_phase=False</em>, <em class="sig-param">detailed_info=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Haploid2DiploidConverter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Haploid2DiploidConverter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Conversion of multiple haploid <cite>.vcf</cite> files to a single <cite>.vcf</cite>.</p>
<dl class="method">
<dt id="medaka.vcf.Haploid2DiploidConverter.meta_info">
<em class="property">property </em><code class="sig-name descname">meta_info</code><a class="headerlink" href="#medaka.vcf.Haploid2DiploidConverter.meta_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the meta information for the combined <cite>.vcf</cite> file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Haploid2DiploidConverter.variants">
<code class="sig-name descname">variants</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Haploid2DiploidConverter.variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Haploid2DiploidConverter.variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Yield diploid variants.</p>
<p>:yields <cite>medaka.vcf.Variant</cite> objs</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.vcf.MetaInfo">
<em class="property">class </em><code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">MetaInfo</code><span class="sig-paren">(</span><em class="sig-param">group</em>, <em class="sig-param">ident</em>, <em class="sig-param">number</em>, <em class="sig-param">typ</em>, <em class="sig-param">descr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#MetaInfo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.MetaInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of a variant file meta data.</p>
</dd></dl>

<dl class="class">
<dt id="medaka.vcf.VCFReader">
<em class="property">class </em><code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">VCFReader</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">cache=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#VCFReader"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.VCFReader" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Basic VCF parser.</p>
<dl class="method">
<dt id="medaka.vcf.VCFReader.fetch">
<code class="sig-name descname">fetch</code><span class="sig-paren">(</span><em class="sig-param">ref_name=None</em>, <em class="sig-param">start=None</em>, <em class="sig-param">end=None</em>, <em class="sig-param">strict=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#VCFReader.fetch"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.VCFReader.fetch" title="Permalink to this definition">¶</a></dt>
<dd><p>Yield all variants spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_name</strong> – reference name (CHROM field of .vcf).</p></li>
<li><p><strong>start</strong> – inclusive start co-ordinate (0-based).</p></li>
<li><p><strong>end</strong> – exclusive end co-ordinate (0-based).</p></li>
<li><p><strong>strict</strong> – if False variants overlapping the region, but not
contained enitrely within the region are yielded also.</p></li>
</ul>
</dd>
<dt class="field-even">Yields</dt>
<dd class="field-even"><p><cite>Variant</cite> instances.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.VCFReader.index">
<code class="sig-name descname">index</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#VCFReader.index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.VCFReader.index" title="Permalink to this definition">¶</a></dt>
<dd><p>Index the input file for faster fetches.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.vcf.VCFWriter">
<em class="property">class </em><code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">VCFWriter</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">mode='w'</em>, <em class="sig-param">header=('CHROM'</em>, <em class="sig-param">'POS'</em>, <em class="sig-param">'ID'</em>, <em class="sig-param">'REF'</em>, <em class="sig-param">'ALT'</em>, <em class="sig-param">'QUAL'</em>, <em class="sig-param">'FILTER'</em>, <em class="sig-param">'INFO'</em>, <em class="sig-param">'FORMAT'</em>, <em class="sig-param">'SAMPLE')</em>, <em class="sig-param">contigs=None</em>, <em class="sig-param">meta_info=None</em>, <em class="sig-param">version='4.1'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#VCFWriter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.VCFWriter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Writing of <cite>Variants</cite> to file.</p>
<dl class="attribute">
<dt id="medaka.vcf.VCFWriter.version_options">
<code class="sig-name descname">version_options</code><em class="property"> = {'4.1', '4.3'}</em><a class="headerlink" href="#medaka.vcf.VCFWriter.version_options" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="medaka.vcf.VCFWriter.write_variant">
<code class="sig-name descname">write_variant</code><span class="sig-paren">(</span><em class="sig-param">variant</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#VCFWriter.write_variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.VCFWriter.write_variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a single variant to file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>variant</strong> – the <cite>Variant</cite> to write.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.VCFWriter.write_variants">
<code class="sig-name descname">write_variants</code><span class="sig-paren">(</span><em class="sig-param">variants</em>, <em class="sig-param">sort=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#VCFWriter.write_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.VCFWriter.write_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Write variants to file, optionally sorting before writing.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="medaka.vcf.Variant">
<em class="property">class </em><code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">Variant</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">pos</em>, <em class="sig-param">ref</em>, <em class="sig-param">alt='.'</em>, <em class="sig-param">ident='.'</em>, <em class="sig-param">qual='.'</em>, <em class="sig-param">filt='.'</em>, <em class="sig-param">info='.'</em>, <em class="sig-param">genotype_data=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Representation of a genomic variant.</p>
<dl class="method">
<dt id="medaka.vcf.Variant.add_tag">
<code class="sig-name descname">add_tag</code><span class="sig-paren">(</span><em class="sig-param">tag</em>, <em class="sig-param">value=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.add_tag"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.add_tag" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a tag (with value).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tag</strong> – tag name.</p></li>
<li><p><strong>value</strong> – tag value.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.alleles">
<em class="property">property </em><code class="sig-name descname">alleles</code><a class="headerlink" href="#medaka.vcf.Variant.alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Return alleles present in genotype.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.deep_copy">
<code class="sig-name descname">deep_copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.deep_copy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.deep_copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the (deep)copy of the <cite>Variant</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.from_text">
<em class="property">classmethod </em><code class="sig-name descname">from_text</code><span class="sig-paren">(</span><em class="sig-param">line</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.from_text"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.from_text" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a <cite>Variant</cite> from a <cite>.vcf</cite> formatted line.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>line</strong> – string representing variant.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.genotype_keys">
<em class="property">property </em><code class="sig-name descname">genotype_keys</code><a class="headerlink" href="#medaka.vcf.Variant.genotype_keys" title="Permalink to this definition">¶</a></dt>
<dd><p>Return genotype format field for writing to`.vcf` file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.genotype_values">
<em class="property">property </em><code class="sig-name descname">genotype_values</code><a class="headerlink" href="#medaka.vcf.Variant.genotype_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genotype data values for writing to <cite>.vcf</cite> file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.get_tag">
<code class="sig-name descname">get_tag</code><span class="sig-paren">(</span><em class="sig-param">tag</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.get_tag"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.get_tag" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the value of a tag by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tag</strong> – tag name.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.gt">
<em class="property">property </em><code class="sig-name descname">gt</code><a class="headerlink" href="#medaka.vcf.Variant.gt" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genotype (or None) for each sample.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.info_string">
<em class="property">property </em><code class="sig-name descname">info_string</code><a class="headerlink" href="#medaka.vcf.Variant.info_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Return info field for writing to <cite>.vcf</cite> file.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.phased">
<em class="property">property </em><code class="sig-name descname">phased</code><a class="headerlink" href="#medaka.vcf.Variant.phased" title="Permalink to this definition">¶</a></dt>
<dd><p>Specify whether variant is phased.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.split_haplotypes">
<code class="sig-name descname">split_haplotypes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.split_haplotypes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.split_haplotypes" title="Permalink to this definition">¶</a></dt>
<dd><p>Split multiploid variants into list of non-ref haploid variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>(haplotype number, <cite>vcf.Variant</cite> or None for ref allele)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.to_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary representation.</p>
</dd></dl>

<dl class="method">
<dt id="medaka.vcf.Variant.trim">
<code class="sig-name descname">trim</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#Variant.trim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.Variant.trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Return new trimmed Variant with minimal ref and alt sequence.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="medaka.vcf.classify_variant">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">classify_variant</code><span class="sig-paren">(</span><em class="sig-param">v</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#classify_variant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.classify_variant" title="Permalink to this definition">¶</a></dt>
<dd><p>Classify a variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>v</strong> – <cite>medaka.vcf.Variant</cite> obj</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>typ: str, variant classification.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.classify_variants">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">classify_variants</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#classify_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.classify_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point to classify variants.</p>
<p>Program to separate variants by class and write to individual files.</p>
<p>Example classifications for subsitutions are ‘snp’ and ‘mnp’, which belong
to the same class-group ‘sub’.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.diploid2haploid">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">diploid2haploid</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#diploid2haploid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.diploid2haploid" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point to split a diploid VCF into two haploid VCFs.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.get_homozygous_regions">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">get_homozygous_regions</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#get_homozygous_regions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.get_homozygous_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point to find homozygous regions from an input diploid VCF.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.haploid2diploid">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">haploid2diploid</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#haploid2diploid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.haploid2diploid" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for merging two haploid vcfs into a diploid vcf.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.parse_string_to_tags">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">parse_string_to_tags</code><span class="sig-paren">(</span><em class="sig-param">string</em>, <em class="sig-param">splitter='</em>, <em class="sig-param">'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#parse_string_to_tags"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.parse_string_to_tags" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a dictionary of “tag” meta data from a string representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>string</strong> – string containing tags.</p></li>
<li><p><strong>splitter</strong> – delimiter of array-valued items.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>dictionary of tags.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.parse_tags_to_string">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">parse_tags_to_string</code><span class="sig-paren">(</span><em class="sig-param">tags</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#parse_tags_to_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.parse_tags_to_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Create string representation of a dictionary of tags.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>tags</strong> – dictionary containing “tag” meta data of a variant.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the string representation of the tags.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.self_return">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">self_return</code><span class="sig-paren">(</span><em class="sig-param">x</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#self_return"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.self_return" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the input.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.split_variants">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">split_variants</code><span class="sig-paren">(</span><em class="sig-param">vcf_fp</em>, <em class="sig-param">trim=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#split_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.split_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Split diploid vcf into two haploid vcfs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vcf</strong> – str, path to input vcf.</p></li>
<li><p><strong>trim</strong> – bool, trim variants to minimal alt and ref and update pos.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>tuple of output files written</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.vcf.vcf2tsv">
<code class="sig-prename descclassname">medaka.vcf.</code><code class="sig-name descname">vcf2tsv</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka/vcf.html#vcf2tsv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.vcf.vcf2tsv" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point to convert vcf to tsv, unpacking info and sample fields.</p>
</dd></dl>

</div>
<div class="section" id="module-medaka">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-medaka" title="Permalink to this headline">¶</a></h2>
<p>Consensus and variant calling of Nanopore sequencing data.</p>
<dl class="function">
<dt id="medaka.check_bgzip_version">
<code class="sig-prename descclassname">medaka.</code><code class="sig-name descname">check_bgzip_version</code><span class="sig-paren">(</span><em class="sig-param">pos=2</em><span class="sig-paren">)</span><a class="headerlink" href="#medaka.check_bgzip_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Get version of an htslib program.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tool</strong> – program name.</p></li>
<li><p><strong>pos</strong> – the position index of the item containing the version
information. e.g. <cite>pos=2</cite> extracts <cite>1.3.1</cite> from
<cite>tabix (htslib) 1.3.1</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the <cite>LooseVersion</cite> number.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.check_htslib_tool_version">
<code class="sig-prename descclassname">medaka.</code><code class="sig-name descname">check_htslib_tool_version</code><span class="sig-paren">(</span><em class="sig-param">tool</em>, <em class="sig-param">pos=2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka.html#check_htslib_tool_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.check_htslib_tool_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Get version of an htslib program.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tool</strong> – program name.</p></li>
<li><p><strong>pos</strong> – the position index of the item containing the version
information. e.g. <cite>pos=2</cite> extracts <cite>1.3.1</cite> from
<cite>tabix (htslib) 1.3.1</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the <cite>LooseVersion</cite> number.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.check_minimap2_version">
<code class="sig-prename descclassname">medaka.</code><code class="sig-name descname">check_minimap2_version</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka.html#check_minimap2_version"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.check_minimap2_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Get minimap2 version.</p>
</dd></dl>

<dl class="function">
<dt id="medaka.check_samtools_version">
<code class="sig-prename descclassname">medaka.</code><code class="sig-name descname">check_samtools_version</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#medaka.check_samtools_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Get version of an htslib program.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tool</strong> – program name.</p></li>
<li><p><strong>pos</strong> – the position index of the item containing the version
information. e.g. <cite>pos=2</cite> extracts <cite>1.3.1</cite> from
<cite>tabix (htslib) 1.3.1</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the <cite>LooseVersion</cite> number.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.check_tabix_version">
<code class="sig-prename descclassname">medaka.</code><code class="sig-name descname">check_tabix_version</code><span class="sig-paren">(</span><em class="sig-param">pos=2</em><span class="sig-paren">)</span><a class="headerlink" href="#medaka.check_tabix_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Get version of an htslib program.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tool</strong> – program name.</p></li>
<li><p><strong>pos</strong> – the position index of the item containing the version
information. e.g. <cite>pos=2</cite> extracts <cite>1.3.1</cite> from
<cite>tabix (htslib) 1.3.1</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>the <cite>LooseVersion</cite> number.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="medaka.report_binaries">
<code class="sig-prename descclassname">medaka.</code><code class="sig-name descname">report_binaries</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/medaka.html#report_binaries"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#medaka.report_binaries" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a report of versions of required programs.</p>
<p>The program will exit with status 1 if a bad version is found.</p>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="future.html" class="btn btn-neutral float-right" title="History and Future Directions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="development.html" class="btn btn-neutral float-left" title="Development" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-19, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>