

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Methylation calling &mdash; Medaka 0.11.2 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Walkthrough" href="walkthrough.html" />
    <link rel="prev" title="Variant calling" href="snp.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Medaka
          

          
          </a>

          
            
            
              <div class="version">
                0.11.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="benchmarks.html">Benchmarks</a></li>
<li class="toctree-l1"><a class="reference internal" href="snp.html">Variant calling</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Methylation calling</a></li>
<li class="toctree-l1"><a class="reference internal" href="walkthrough.html">Walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="draft_origin.html">Origin of the draft sequence</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="future.html">History and Future Directions</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Medaka</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Methylation calling</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/methylation.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="methylation-calling">
<h1>Methylation calling<a class="headerlink" href="#methylation-calling" title="Permalink to this headline">¶</a></h1>
<p>Medaka includes a basic workflow for aggregating Guppy basecalling results
for Dcm, Dam, and CpG methylation. The workflow is currently very preliminary
and subject to change and improvement.</p>
<p>Aggregating the information from Guppy outputs is a two stage process, first
the basecalling results are extracted <cite>.fast5</cite> files and placed in a <cite>.bam</cite>
file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">FAST5PATH</span><span class="o">=</span>guppy/workspace
<span class="nv">REFERENCE</span><span class="o">=</span>grch38.fasta
<span class="nv">OUTBAM</span><span class="o">=</span>meth.bam
medaka methylation guppy2sam <span class="si">${</span><span class="nv">FAST5PATH</span><span class="si">}</span> <span class="si">${</span><span class="nv">REFERENCE</span><span class="si">}</span> <span class="se">\</span>
    --workers <span class="m">74</span> --recursive <span class="se">\</span>
    <span class="p">|</span> samtools sort -@ <span class="m">8</span> <span class="p">|</span> samtools view -b -@ <span class="m">8</span> &gt; <span class="si">${</span><span class="nv">OUTBAM</span><span class="si">}</span>
samtools index <span class="si">${</span><span class="nv">OUTBAM</span><span class="si">}</span>
</pre></div>
</div>
<p>This program will extract both the basecall sequence and methylation scores,
align the basecall to the reference, and store results in a standard format.
In this preliminary workflow the methylation scores are stored in two SAM
tags, ‘MC’ and ‘MA’, one each for 5mC and 6mA respectively. The tags are
8bit integer array-values, one value per basecall position. This is a
different form to that proposed in the current
<a class="reference external" href="https://github.com/samtools/hts-specs/pull/418/files">hts-specs proposition</a>,
but allows for more trivial parsing.</p>
<p>The second step is to aggregate the reference-aligned information to produce
a simple tabular summary of read methylation counts:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">BAM</span><span class="o">=</span>meth.bam
<span class="nv">REFERENCE</span><span class="o">=</span>grch38.fasta
<span class="nv">REGION</span><span class="o">=</span>chr20:500000-1000000
<span class="nv">OUTPUT</span><span class="o">=</span>meth.tsv
medaka methylation call --meth cpg <span class="si">${</span><span class="nv">BAM</span><span class="si">}</span> <span class="si">${</span><span class="nv">REFERENCE</span><span class="si">}</span> <span class="si">${</span><span class="nv">REGION</span><span class="si">}</span> <span class="si">${</span><span class="nv">OUTPUT</span><span class="si">}</span>
</pre></div>
</div>
<p>Here the option <code class="docutils literal notranslate"><span class="pre">--meth</span> <span class="pre">cpg</span></code> indicates that loci containing the sequence
motif <code class="docutils literal notranslate"><span class="pre">CG</span></code> should be examined for 5mC presence. Other choices are
<code class="docutils literal notranslate"><span class="pre">dcm</span></code> (motifs <code class="docutils literal notranslate"><span class="pre">CCAGG</span></code> and <code class="docutils literal notranslate"><span class="pre">CCTGG</span></code>) for 5mC and <code class="docutils literal notranslate"><span class="pre">dam</span></code> (<code class="docutils literal notranslate"><span class="pre">GATC</span></code>) for 6mA.</p>
<p>The output file is a simple tab-delimited text file with columns:
‘ref.name’, ‘position’, ‘motif’, ‘fwd.meth.count’, ‘rev.meth.count’,
‘fwd.canon.count’, and ‘rev.canon.count’. Here fwd./rev. indicate counts on the
two DNA strands and meth./canon. indicate counts for methylated and
canonical bases. Note that the position field records the position of the
first base in the motif recorded.</p>
<p>As a first demonstration of this method the plot below shows a comparison with
bisulphite sequencing <a class="reference external" href="https://www.encodeproject.org/files/ENCFF835NTC/">results</a>
for chromosome 20 of the NA12878 sample.</p>
<img alt="_images/methcalls.png" class="align-center" src="_images/methcalls.png" />
<p>Reasonable correlation is found, especially given the simplicity of the aggregation
method. A slight bias toward lower methylation frequency is observed through the range,
except at very low frequencies where the method has a bias upward.</p>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="walkthrough.html" class="btn btn-neutral float-right" title="Walkthrough" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="snp.html" class="btn btn-neutral float-left" title="Variant calling" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017-19, Oxford Nanopore Technologies

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>